{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Processing, Clustering and Visualising in MDV the TAURUS data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Importing the required packages for data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Importing the required packages for MDV set up and visualisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from mdvtools.mdvproject import MDVProject\n",
    "from mdvtools.charts.dot_plot import DotPlot\n",
    "from mdvtools.charts.scatter_plot import ScatterPlot\n",
    "from mdvtools.charts.box_plot import BoxPlot\n",
    "from mdvtools.charts.stacked_row_plot import StackedRowChart"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data analysis section"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.10.1 anndata==0.9.2 umap==0.5.6 numpy==1.26.4 scipy==1.13.0 pandas==2.2.2 scikit-learn==1.4.2 statsmodels==0.14.2 igraph==0.11.4 pynndescent==0.5.12\n"
     ]
    }
   ],
   "source": [
    "# scanpy parameters for feedback level setting\n",
    "sc.settings.verbosity = 3  # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_header() # printing a header of introductory information about the environment/library used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "__init__.py         csv_to_static.py    \u001b[34mquerytools\u001b[m\u001b[m          \u001b[34mtest_projects\u001b[m\u001b[m\n",
      "\u001b[34m__pycache__\u001b[m\u001b[m         \u001b[34mdbutils\u001b[m\u001b[m             requirements.txt    \u001b[34mtests\u001b[m\u001b[m\n",
      "chart_prototypes.py \u001b[34mllm\u001b[m\u001b[m                 server.py           websocket.py\n",
      "\u001b[34mcharts\u001b[m\u001b[m              mdv_desktop.py      spatial.py\n",
      "chat.py             mdvproject.py       \u001b[34mstatic\u001b[m\u001b[m\n",
      "conversions.py      project_router.py   \u001b[34mtemplates\u001b[m\u001b[m\n"
     ]
    }
   ],
   "source": [
    "! ls ../"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "bcell_data = sc.read_h5ad(\"../llm/sample_data/bcell_viz_ready_revised.h5ad\",)\n",
    "\n",
    "# bcell_data = sc.read_h5ad('../../../../../../mariak/anndata_obj/sub_buckets/bcells_final.h5ad')\n",
    "# cd4tcell_data = sc.read_h5ad('../../../../../../mariak/anndata_obj/sub_buckets/cd4tcells_final.h5ad')\n",
    "# cd8tcell_data = sc.read_h5ad('../../../../../../mariak/anndata_obj/sub_buckets/cd8tcells_final.h5ad')\n",
    "# epithelium_colonic_data = sc.read_h5ad('../../../../../../mariak/anndata_obj/sub_buckets/epicolonic_final.h5ad')\n",
    "# endothelial_data = sc.read_h5ad('../../../../../../mariak/anndata_obj/sub_buckets/vasc_final.h5ad') \n",
    "# epi_ileal_data = sc.read_h5ad('../../../../../../mariak/anndata_obj/sub_buckets/ilealepi_final.h5ad')\n",
    "# fib_peri_data = sc.read_h5ad('../../../../../../mariak/anndata_obj/sub_buckets/fibperi_final.h5ad')\n",
    "# ilc_data = sc.read_h5ad('../../../../../../mariak/anndata_obj/sub_buckets/ilc_final.h5ad') \n",
    "# myeloid_data = sc.read_h5ad('../../../../../../mariak/anndata_obj/sub_buckets/myeloid_final.h5ad') \n",
    "# plasmacell_data = sc.read_h5ad('../../../../../../mariak/anndata_obj/sub_buckets/plasmacells_final.h5ad') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sc.pp.subsample(bcell_data, 0.05)\n",
    "# sc.pp.subsample(cd4tcell_data, 0.05)\n",
    "# sc.pp.subsample(cd8tcell_data, 0.05)\n",
    "# sc.pp.subsample(epithelium_colonic_data, 0.05)\n",
    "# sc.pp.subsample(endothelial_data, 0.05)\n",
    "# sc.pp.subsample(epi_ileal_data, 0.05)\n",
    "# sc.pp.subsample(fib_peri_data, 0.05)\n",
    "# sc.pp.subsample(ilc_data, 0.05)\n",
    "# sc.pp.subsample(myeloid_data, 0.05)\n",
    "# sc.pp.subsample(plasmacell_data, 0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# B cells\n",
    "# Saving count data\n",
    "bcell_data.layers[\"counts\"] = bcell_data.X.copy()\n",
    "# total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "#sc.pp.normalize_total(bcell_data, target_sum=1e4)\n",
    "# logarithmising the data\n",
    "#sc.pp.log1p(bcell_data)\n",
    "\n",
    "# # # CD4T cells\n",
    "# # # Saving count data\n",
    "# cd4tcell_data.layers[\"counts\"] = cd4tcell_data.X.copy()\n",
    "# # # total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "# # sc.pp.normalize_total(cd4tcell_data, target_sum=1e4)\n",
    "# # # logarithmising the data\n",
    "# # sc.pp.log1p(cd4tcell_data)\n",
    "\n",
    "# # # CD8T cells\n",
    "# # # Saving count data\n",
    "# cd8tcell_data.layers[\"counts\"] = cd8tcell_data.X.copy()\n",
    "# # # total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "# # sc.pp.normalize_total(cd8tcell_data, target_sum=1e4)\n",
    "# # # logarithmising the data\n",
    "# # sc.pp.log1p(cd8tcell_data)\n",
    "\n",
    "# # # Endothelium cells\n",
    "# # # Saving count data\n",
    "# endothelial_data.layers[\"counts\"] = endothelial_data.X.copy()\n",
    "# # # total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "# # sc.pp.normalize_total(endothelial_data, target_sum=1e4)\n",
    "# # # logarithmising the data\n",
    "# # sc.pp.log1p(endothelial_data)\n",
    "\n",
    "# # # epithelial cells\n",
    "# # # Saving count data\n",
    "# epi_ileal_data.layers[\"counts\"] = epi_ileal_data.X.copy()\n",
    "# # # total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "# # sc.pp.normalize_total(epi_ileal_data, target_sum=1e4)\n",
    "# # # logarithmising the data\n",
    "# # sc.pp.log1p(epi_ileal_data)\n",
    "\n",
    "# # # Epithelium colonic cells\n",
    "# # # Saving count data\n",
    "# epithelium_colonic_data.layers[\"counts\"] = epithelium_colonic_data.X.copy()\n",
    "# # # total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "# # sc.pp.normalize_total(epithelium_colonic_data, target_sum=1e4)\n",
    "# # # logarithmising the data\n",
    "# # sc.pp.log1p(epithelium_colonic_data)\n",
    "\n",
    "# # # Fiber cells\n",
    "# # # Saving count data\n",
    "# fib_peri_data.layers[\"counts\"] = fib_peri_data.X.copy()\n",
    "# # # total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "# # sc.pp.normalize_total(fib_peri_data, target_sum=1e4)\n",
    "# # # logarithmising the data\n",
    "# # sc.pp.log1p(fib_peri_data)\n",
    "\n",
    "# # # Ilc cells\n",
    "# # # Saving count data\n",
    "# ilc_data.layers[\"counts\"] = ilc_data.X.copy()\n",
    "# # # total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "# # sc.pp.normalize_total(ilc_data, target_sum=1e4)\n",
    "# # # logarithmising the data\n",
    "# # sc.pp.log1p(ilc_data)\n",
    "\n",
    "# # # Myeloid cells\n",
    "# # # Saving count data\n",
    "# myeloid_data.layers[\"counts\"] = myeloid_data.X.copy()\n",
    "# # # total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "# # sc.pp.normalize_total(myeloid_data, target_sum=1e4)\n",
    "# # # logarithmising the data\n",
    "# # sc.pp.log1p(myeloid_data)\n",
    "\n",
    "# # # Plasma cells\n",
    "# # # Saving count data\n",
    "# plasmacell_data.layers[\"counts\"] = plasmacell_data.X.copy()\n",
    "# # # total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "# # sc.pp.normalize_total(plasmacell_data, target_sum=1e4)\n",
    "# # # logarithmising the data\n",
    "# # sc.pp.log1p(plasmacell_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## B CELLS\n",
    "# cells dataframe \n",
    "cells_df_bcell = pd.DataFrame(bcell_data.obs)\n",
    "\n",
    "# adding the umap data to the cells dataframe\n",
    "cells_df_bcell[\"UMAP 1\"] = np.array(bcell_data.obsm[\"X_umap\"])[:, 0]\n",
    "cells_df_bcell[\"UMAP 2\"] = np.array(bcell_data.obsm[\"X_umap\"])[:, 1]\n",
    "cells_df_bcell[\"Cell ID\"] = bcell_data.obs.index\n",
    "cells_df_bcell.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "                         \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "#cells_df_bcell = cells_df_bcell.iloc[:, [0,1,2,3,4,6,10,33,37,38,39,40]]\n",
    "\n",
    "# genes dataframe\n",
    "gene_table_bcell = bcell_data.var\n",
    "gene_table_bcell[\"gene_id\"]=gene_table_bcell.index\n",
    "\n",
    "\n",
    "# # ## CD4T CELLS\n",
    "# # # cells dataframe \n",
    "# cells_df_cd4t = pd.DataFrame(cd4tcell_data.obs)\n",
    "\n",
    "# # # adding the umap data to the cells dataframe\n",
    "# # #umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "# cells_df_cd4t[\"UMAP 1\"] = np.array(cd4tcell_data.obsm[\"X_umap\"])[:, 0]\n",
    "# cells_df_cd4t[\"UMAP 2\"] = np.array(cd4tcell_data.obsm[\"X_umap\"])[:, 1]\n",
    "# cells_df_cd4t[\"Cell ID\"] = cd4tcell_data.obs.index\n",
    "# cells_df_cd4t.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "#                          \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "# cells_df_cd4t = cells_df_cd4t.iloc[:, [0,1,2,3,4,6,10,33,37,38,39,40]]\n",
    "\n",
    "# # genes dataframe\n",
    "# gene_table_cd4t = cd4tcell_data.var\n",
    "# gene_table_cd4t[\"gene_id\"]=gene_table_cd4t.index\n",
    "\n",
    "# ## CD8T CELLS\n",
    "# # cells dataframe \n",
    "# cells_df_cd8t = pd.DataFrame(cd8tcell_data.obs)\n",
    "\n",
    "# # adding the umap data to the cells dataframe\n",
    "# #umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "# cells_df_cd8t[\"UMAP 1\"] = np.array(cd8tcell_data.obsm[\"X_umap\"])[:, 0]\n",
    "# cells_df_cd8t[\"UMAP 2\"] = np.array(cd8tcell_data.obsm[\"X_umap\"])[:, 1]\n",
    "# cells_df_cd8t[\"Cell ID\"] = cd8tcell_data.obs.index\n",
    "# cells_df_cd8t.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "#                          \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "# cells_df_cd8t = cells_df_cd8t.iloc[:, [0,1,2,3,4,6,10,33,37,38,39,40]]\n",
    "\n",
    "# # genes dataframe\n",
    "# gene_table_cd8t = cd8tcell_data.var\n",
    "# gene_table_cd8t[\"gene_id\"]=gene_table_cd8t.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## COLONIC EPITHELIAL CELLS\n",
    "# # cells dataframe \n",
    "# cells_df_col_epi = pd.DataFrame(epithelium_colonic_data.obs)\n",
    "\n",
    "# # adding the umap data to the cells dataframe\n",
    "# #umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "# cells_df_col_epi[\"UMAP 1\"] = np.array(epithelium_colonic_data.obsm[\"X_umap\"])[:, 0]\n",
    "# cells_df_col_epi[\"UMAP 2\"] = np.array(epithelium_colonic_data.obsm[\"X_umap\"])[:, 1]\n",
    "# cells_df_col_epi[\"Cell ID\"] = epithelium_colonic_data.obs.index\n",
    "# cells_df_col_epi.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "#                          \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "# cells_df_col_epi = cells_df_col_epi.iloc[:, [0,1,2,3,4,6,10,33,37,38,39,40]]\n",
    "\n",
    "# # genes dataframe\n",
    "# gene_table_col_epi = epithelium_colonic_data.var\n",
    "# gene_table_col_epi[\"gene_id\"]=gene_table_col_epi.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## ENDOTHELIAL CELLS\n",
    "# # cells dataframe \n",
    "# cells_df_endo = pd.DataFrame(endothelial_data.obs)\n",
    "\n",
    "# # adding the umap data to the cells dataframe\n",
    "# #umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "# cells_df_endo[\"UMAP 1\"] = np.array(endothelial_data.obsm[\"X_umap\"])[:, 0]\n",
    "# cells_df_endo[\"UMAP 2\"] = np.array(endothelial_data.obsm[\"X_umap\"])[:, 1]\n",
    "# cells_df_endo[\"Cell ID\"] = endothelial_data.obs.index\n",
    "# cells_df_endo.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "#                          \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "# cells_df_endo = cells_df_endo.iloc[:, [0,1,2,3,4,6,10,33,37,38,39,40]]\n",
    "\n",
    "# # genes dataframe\n",
    "# gene_table_endo = endothelial_data.var\n",
    "# gene_table_endo[\"gene_id\"]=gene_table_endo.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## FIBROBLAST CELLS\n",
    "# # cells dataframe \n",
    "# cells_df_fib = pd.DataFrame(fib_peri_data.obs)\n",
    "\n",
    "# # adding the umap data to the cells dataframe\n",
    "# #umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "# cells_df_fib[\"UMAP 1\"] = np.array(fib_peri_data.obsm[\"X_umap\"])[:, 0]\n",
    "# cells_df_fib[\"UMAP 2\"] = np.array(fib_peri_data.obsm[\"X_umap\"])[:, 1]\n",
    "# cells_df_fib[\"Cell ID\"] = fib_peri_data.obs.index\n",
    "# cells_df_fib.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "#                          \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "# cells_df_fib = cells_df_fib.iloc[:, [0,1,2,3,4,6,10,33,37,38,39,40]]\n",
    "\n",
    "# # genes dataframe\n",
    "# gene_table_fib = fib_peri_data.var\n",
    "# gene_table_fib[\"gene_id\"]=gene_table_fib.index\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## ILEAL EPITHELIAL CELLS\n",
    "# # cells dataframe \n",
    "# cells_df_ileal = pd.DataFrame(epi_ileal_data.obs)\n",
    "\n",
    "# # adding the umap data to the cells dataframe\n",
    "# #umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "# cells_df_ileal[\"UMAP 1\"] = np.array(epi_ileal_data.obsm[\"X_umap\"])[:, 0]\n",
    "# cells_df_ileal[\"UMAP 2\"] = np.array(epi_ileal_data.obsm[\"X_umap\"])[:, 1]\n",
    "# cells_df_ileal[\"Cell ID\"] = epi_ileal_data.obs.index\n",
    "# cells_df_ileal.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "#                          \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "# cells_df_ileal = cells_df_ileal.iloc[:, [0,1,2,3,4,6,10,33,37,38,39,40]]\n",
    "# # genes dataframe\n",
    "# gene_table_ileal = epi_ileal_data.var\n",
    "# gene_table_ileal[\"gene_id\"]=gene_table_ileal.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## INNATE LYMPHOID CELLS\n",
    "# # cells dataframe \n",
    "# cells_df_ilc = pd.DataFrame(ilc_data.obs)\n",
    "\n",
    "# # adding the umap data to the cells dataframe\n",
    "# #umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "# cells_df_ilc[\"UMAP 1\"] = np.array(ilc_data.obsm[\"X_umap\"])[:, 0]\n",
    "# cells_df_ilc[\"UMAP 2\"] = np.array(ilc_data.obsm[\"X_umap\"])[:, 1]\n",
    "# cells_df_ilc[\"Cell ID\"] = ilc_data.obs.index\n",
    "# cells_df_ilc.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "#                          \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "# cells_df_ilc = cells_df_ilc.iloc[:, [0,1,2,3,4,6,10,33,37,38,39,40]]\n",
    "\n",
    "# # genes dataframe\n",
    "# gene_table_ilc = ilc_data.var\n",
    "# gene_table_ilc[\"gene_id\"]=gene_table_ilc.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## MYELOID CELLS\n",
    "# # cells dataframe \n",
    "# cells_df_myel = pd.DataFrame(myeloid_data.obs)\n",
    "\n",
    "# # adding the umap data to the cells dataframe\n",
    "# #umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "# cells_df_myel[\"UMAP 1\"] = np.array(myeloid_data.obsm[\"X_umap\"])[:, 0]\n",
    "# cells_df_myel[\"UMAP 2\"] = np.array(myeloid_data.obsm[\"X_umap\"])[:, 1]\n",
    "# cells_df_myel[\"Cell ID\"] = myeloid_data.obs.index\n",
    "# cells_df_myel.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "#                          \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "# cells_df_myel = cells_df_myel.iloc[:, [0,1,2,3,4,6,10,33,37,38,39,40]]\n",
    "\n",
    "# # genes dataframe\n",
    "# gene_table_myel = myeloid_data.var\n",
    "# gene_table_myel[\"gene_id\"]=gene_table_myel.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## PLASMA CELLS\n",
    "# # cells dataframe \n",
    "# cells_df_plasma = pd.DataFrame(plasmacell_data.obs)\n",
    "\n",
    "# # adding the umap data to the cells dataframe\n",
    "# #umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "# cells_df_plasma[\"UMAP 1\"] = np.array(plasmacell_data.obsm[\"X_umap\"])[:, 0]\n",
    "# cells_df_plasma[\"UMAP 2\"] = np.array(plasmacell_data.obsm[\"X_umap\"])[:, 1]\n",
    "# cells_df_plasma[\"Cell ID\"] = plasmacell_data.obs.index\n",
    "# cells_df_plasma.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "#                          \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "# cells_df_plasma = cells_df_plasma.iloc[:, [0,1,2,3,4,6,10,33,37,38,39,40]]\n",
    "\n",
    "# # genes dataframe\n",
    "# gene_table_plasma = plasmacell_data.var\n",
    "# gene_table_plasma[\"gene_id\"]=gene_table_plasma.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## B CELLS\n",
    "\n",
    "# Cells charts\n",
    "# creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "stacked_row_chart_bcell = StackedRowChart(\n",
    "    title=\"Abundance plot\",\n",
    "    params=[\"Patient\", \"Cell state\"],\n",
    "    position=[930, 10],\n",
    "    size=[580, 450]\n",
    ")\n",
    "# configuring the row chart\n",
    "stacked_row_chart_bcell.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "#stacked_row_chart_bcell.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "\n",
    "#creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "gene_name = \"HSPE1\"\n",
    "\n",
    "dot_plot_bcell = DotPlot(\n",
    "    title= f\"Percent of gene expression per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_bcell.index.get_loc(gene_name)}\"],\n",
    "    #params = [\"Cell state\",{\"linkedDsName\":\"B genes\",\"maxItems\":10,\"type\":\"RowsAsColsQuery\"}], \n",
    "    size=[450, 300],\n",
    "    position=[10, 470]\n",
    ")\n",
    "\n",
    "# configuring the dot plot\n",
    "dot_plot_bcell.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_bcell.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_bcell.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_bcell.set_color_scale(log_scale=False)\n",
    "dot_plot_bcell.set_color_legend(True, [40, 10])\n",
    "dot_plot_bcell.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "box_plot_bcell = BoxPlot(\n",
    "    title=f\"Gene expression per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_bcell.index.get_loc(gene_name)}\"],\n",
    "    #params = [\"Cell state\", {\"linkedDsName\":\"B genes\",\"maxItems\":1,\"type\":\"RowsAsColsQuery\"}],\n",
    "    size=[650, 300],\n",
    "    position=[470, 470]\n",
    ")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot_bcell = ScatterPlot(\n",
    "    title=\"B cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[10, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "# configuring the scatter plot\n",
    "scatter_plot_bcell.set_color_by(\"Cell state\")\n",
    "\n",
    "\n",
    "from mdvtools.charts.scatter_plot import ScatterPlot\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot2_bcell = ScatterPlot(\n",
    "    title=\"B cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[470, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    "    #color_by= {\"linkedDsName\":\"B genes\",\"maxItems\":1,\"type\":\"RowsAsColsQuery\"}\n",
    ")\n",
    "\n",
    "# configuring the scatter plot\n",
    "#scatter_plot2_bcell.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_bcell.index.get_loc(gene_name)}\")\n",
    "#scatter_plot2_bcell.set_color_by{\"linkedDsName\":\"genes\",\"maxItems\":1,\"type\":\"RowsAsColsQuery\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## cd8T CELLS\n",
    "\n",
    "# # Cells charts\n",
    "# # creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "# stacked_row_chart_cd8t = StackedRowChart(\n",
    "#     title=\"Abundance plot\",\n",
    "#     params=[\"Patient\", \"Cell state\"],\n",
    "#     position=[930, 10],\n",
    "#     size=[580, 450]\n",
    "# )\n",
    "# # configuring the row chart\n",
    "# stacked_row_chart_cd8t.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "# #stacked_row_chart_cd8t.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "\n",
    "# #creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "# gene_name = \"HSPE1\"\n",
    "\n",
    "# dot_plot_cd8t = DotPlot(\n",
    "#     title= f\"Percent of gene expression per cell state\",\n",
    "#     #params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_cd8t.index.get_loc(gene_name)}\"],\n",
    "#     params = [\"Cell state\",{\"linkedDsName\":\"CD8T genes\",\"maxItems\":10,\"type\":\"RowsAsColsQuery\"}],\n",
    "#     size=[450, 300],\n",
    "#     position=[10, 470]\n",
    "# )\n",
    "\n",
    "# # configuring the dot plot\n",
    "# dot_plot_cd8t.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_cd8t.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_cd8t.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_cd8t.set_color_scale(log_scale=False)\n",
    "# dot_plot_cd8t.set_color_legend(True, [40, 10])\n",
    "# dot_plot_cd8t.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "# box_plot_cd8t = BoxPlot(\n",
    "#     title=f\"Gene expression per cell state\",\n",
    "#     #params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_cd8t.index.get_loc(gene_name)}\"],\n",
    "#     params = [\"Cell state\", {\"linkedDsName\":\"CD8T genes\",\"maxItems\":1,\"type\":\"RowsAsColsQuery\"}],\n",
    "#     size=[650, 300],\n",
    "#     position=[470, 470]\n",
    "# )\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot_cd8t = ScatterPlot(\n",
    "#     title=\"CD8+ T innate T NK cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[10, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot_cd8t.set_color_by(\"Cell state\")\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot2_cd8t = ScatterPlot(\n",
    "#     title=\"CD8+ T innate T NK cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[470, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot2_cd8t.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_cd8t.index.get_loc(gene_name)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## CD4T CELLS\n",
    "\n",
    "# # Cells charts\n",
    "# # creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "# stacked_row_chart_cd4t = StackedRowChart(\n",
    "#     title=\"Abundance plot\",\n",
    "#     params=[\"Patient\", \"Cell state\"],\n",
    "#     position=[930, 10],\n",
    "#     size=[580, 450]\n",
    "# )\n",
    "# # configuring the row chart\n",
    "# stacked_row_chart_cd4t.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "# #stacked_row_chart_cd4t.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "# #creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "# gene_name = \"HSPE1\"\n",
    "\n",
    "# dot_plot_cd4t = DotPlot(\n",
    "#     title= f\"Percent of gene expression per cell state\",\n",
    "#     #params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_cd4t.index.get_loc(gene_name)}\"],\n",
    "#     params = [\"Cell state\",{\"linkedDsName\":\"CD4T genes\",\"maxItems\":10,\"type\":\"RowsAsColsQuery\"}],\n",
    "#     size=[450, 300],\n",
    "#     position=[10, 470]\n",
    "# )\n",
    "\n",
    "# # configuring the dot plot\n",
    "# dot_plot_cd4t.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_cd4t.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_cd4t.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_cd4t.set_color_scale(log_scale=False)\n",
    "# dot_plot_cd4t.set_color_legend(True, [40, 10])\n",
    "# dot_plot_cd4t.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "# box_plot_cd4t = BoxPlot(\n",
    "#     title=f\"Gene expression per cell state\",\n",
    "#     #params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_cd4t.index.get_loc(gene_name)}\"],\n",
    "#     params = [\"Cell state\", {\"linkedDsName\":\"CD4T genes\",\"maxItems\":1,\"type\":\"RowsAsColsQuery\"}],\n",
    "#     size=[650, 300],\n",
    "#     position=[470, 470]\n",
    "# )\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot_cd4t = ScatterPlot(\n",
    "#     title=\"CD4+ T cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[10, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot_cd4t.set_color_by(\"Cell state\")\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot2_cd4t = ScatterPlot(\n",
    "#     title=\"CD4+ T cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[470, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot2_cd4t.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_cd4t.index.get_loc(gene_name)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## COLONIC EPITHELIAL CELLS\n",
    "\n",
    "# # Cells charts\n",
    "# # creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "# stacked_row_chart_col_epi = StackedRowChart(\n",
    "#     title=\"Abundance plot\",\n",
    "#     params=[\"Patient\", \"Cell state\"],\n",
    "#     position=[930, 10],\n",
    "#     size=[580, 450]\n",
    "# )\n",
    "# # configuring the row chart\n",
    "# stacked_row_chart_col_epi.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "# #stacked_row_chart_col_epi.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "\n",
    "# #creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "# gene_name = \"HSPE1\"\n",
    "\n",
    "# dot_plot_col_epi = DotPlot(\n",
    "#     title= f\"Percent of gene expression per cell state\",\n",
    "#     #params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_col_epi.index.get_loc(gene_name)}\"],\n",
    "#     params = [\"Cell state\",{\"linkedDsName\":\"Colonic epithelial genes\",\"maxItems\":10,\"type\":\"RowsAsColsQuery\"}],\n",
    "#     size=[450, 300],\n",
    "#     position=[10, 470]\n",
    "# )\n",
    "\n",
    "# # configuring the dot plot\n",
    "# dot_plot_col_epi.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_col_epi.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_col_epi.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_col_epi.set_color_scale(log_scale=False)\n",
    "# dot_plot_col_epi.set_color_legend(True, [40, 10])\n",
    "# dot_plot_col_epi.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "# box_plot_col_epi = BoxPlot(\n",
    "#     title=f\"Gene expression per cell state\",\n",
    "#     #params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_col_epi.index.get_loc(gene_name)}\"],\n",
    "#     params = [\"Cell state\", {\"linkedDsName\":\"Colonic epithelial genes\",\"maxItems\":1,\"type\":\"RowsAsColsQuery\"}],\n",
    "#     size=[650, 300],\n",
    "#     position=[470, 470]\n",
    "# )\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot_col_epi = ScatterPlot(\n",
    "#     title=\"Colonic epithelial cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[10, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot_col_epi.set_color_by(\"Cell state\")\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot2_col_epi = ScatterPlot(\n",
    "#     title=\"Colonic epithelial cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[470, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot2_col_epi.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_col_epi.index.get_loc(gene_name)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## ENDOTHELIAL CELLS\n",
    "\n",
    "# # Cells charts\n",
    "# # creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "# stacked_row_chart_endo = StackedRowChart(\n",
    "#     title=\"Abundance plot\",\n",
    "#     params=[\"Patient\", \"Cell state\"],\n",
    "#     position=[930, 10],\n",
    "#     size=[580, 450]\n",
    "# )\n",
    "# # configuring the row chart\n",
    "# stacked_row_chart_endo.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "# #stacked_row_chart_endo.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "# #creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "# gene_name = \"HSPE1\"\n",
    "\n",
    "# dot_plot_endo = DotPlot(\n",
    "#     title= f\"Percent of gene expression per cell state\",\n",
    "#     #params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_endo.index.get_loc(gene_name)}\"],\n",
    "#     params = [\"Cell state\",{\"linkedDsName\":\"Endothelial genes\",\"maxItems\":10,\"type\":\"RowsAsColsQuery\"}],\n",
    "#     size=[450, 300],\n",
    "#     position=[10, 470]\n",
    "# )\n",
    "\n",
    "# # configuring the dot plot\n",
    "# dot_plot_endo.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_endo.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_endo.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_endo.set_color_scale(log_scale=False)\n",
    "# dot_plot_endo.set_color_legend(True, [40, 10])\n",
    "# dot_plot_endo.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "# box_plot_endo = BoxPlot(\n",
    "#     title=f\"Gene expression per cell state\",\n",
    "#     #params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_endo.index.get_loc(gene_name)}\"],\n",
    "#     params = [\"Cell state\", {\"linkedDsName\":\"Endothelial genes\",\"maxItems\":1,\"type\":\"RowsAsColsQuery\"}],\n",
    "#     size=[650, 300],\n",
    "#     position=[470, 470]\n",
    "# )\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot_endo = ScatterPlot(\n",
    "#     title=\"Endothelial cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[10, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot_endo.set_color_by(\"Cell state\")\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot2_endo = ScatterPlot(\n",
    "#     title=\"Endothelial cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[470, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot2_endo.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_endo.index.get_loc(gene_name)}\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## FIBROBLAST PERICYTE CELLS\n",
    "\n",
    "# # Cells charts\n",
    "# # creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "# stacked_row_chart_fib = StackedRowChart(\n",
    "#     title=\"Abundance plot\",\n",
    "#     params=[\"Patient\", \"Cell state\"],\n",
    "#     position=[930, 10],\n",
    "#     size=[580, 450]\n",
    "# )\n",
    "# # configuring the row chart\n",
    "# stacked_row_chart_fib.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "# #stacked_row_chart_fib.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "# #creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "# gene_name = \"HSPE1\"\n",
    "\n",
    "# dot_plot_fib = DotPlot(\n",
    "#     title= f\"Percent of gene expression per cell state\",\n",
    "#     #params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_fib.index.get_loc(gene_name)}\"],\n",
    "#     params = [\"Cell state\",{\"linkedDsName\":\"Fibroblast pericyte genes\",\"maxItems\":10,\"type\":\"RowsAsColsQuery\"}],\n",
    "#     size=[450, 300],\n",
    "#     position=[10, 470]\n",
    "# )\n",
    "\n",
    "# # configuring the dot plot\n",
    "# dot_plot_fib.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_fib.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_fib.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_fib.set_color_scale(log_scale=False)\n",
    "# dot_plot_fib.set_color_legend(True, [40, 10])\n",
    "# dot_plot_fib.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "# box_plot_fib = BoxPlot(\n",
    "#     title=f\"Gene expression per cell state\",\n",
    "#     #params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_fib.index.get_loc(gene_name)}\"],\n",
    "#     params = [\"Cell state\", {\"linkedDsName\":\"Fibroblast pericyte genes\",\"maxItems\":1,\"type\":\"RowsAsColsQuery\"}],\n",
    "#     size=[650, 300],\n",
    "#     position=[470, 470]\n",
    "# )\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot_fib = ScatterPlot(\n",
    "#     title=\"Fibroblast pericyte cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[10, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot_fib.set_color_by(\"Cell state\")\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot2_fib = ScatterPlot(\n",
    "#     title=\"Fibroblast pericyte cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[470, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot2_fib.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_fib.index.get_loc(gene_name)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## ILEAL EPITHELIAL CELLS\n",
    "\n",
    "# # Cells charts\n",
    "# # creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "# stacked_row_chart_ileal = StackedRowChart(\n",
    "#     title=\"Abundance plot\",\n",
    "#     params=[\"Patient\", \"Cell state\"],\n",
    "#     position=[930, 10],\n",
    "#     size=[580, 450]\n",
    "# )\n",
    "# # configuring the row chart\n",
    "# stacked_row_chart_ileal.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "# #stacked_row_chart_ileal.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "# #creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "# gene_name = \"HSPE1\"\n",
    "\n",
    "# dot_plot_ileal = DotPlot(\n",
    "#     title= f\"Percent of gene expression per cell state\",\n",
    "#     #params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_ileal.index.get_loc(gene_name)}\"],\n",
    "#     params = [\"Cell state\",{\"linkedDsName\":\"Ileal epithelial genes\",\"maxItems\":10,\"type\":\"RowsAsColsQuery\"}],\n",
    "#     size=[450, 300],\n",
    "#     position=[10, 470]\n",
    "# )\n",
    "\n",
    "# # configuring the dot plot\n",
    "# dot_plot_ileal.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_ileal.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_ileal.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_ileal.set_color_scale(log_scale=False)\n",
    "# dot_plot_ileal.set_color_legend(True, [40, 10])\n",
    "# dot_plot_ileal.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "# box_plot_ileal = BoxPlot(\n",
    "#     title=f\"Gene expression per cell state\",\n",
    "#     #params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_ileal.index.get_loc(gene_name)}\"],\n",
    "#     params = [\"Cell state\", {\"linkedDsName\":\"Ileal epithelial genes\",\"maxItems\":1,\"type\":\"RowsAsColsQuery\"}],\n",
    "#     size=[650, 300],\n",
    "#     position=[470, 470]\n",
    "# )\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot_ileal = ScatterPlot(\n",
    "#     title=\"Ileal epithelial cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[10, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot_ileal.set_color_by(\"Cell state\")\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot2_ileal = ScatterPlot(\n",
    "#     title=\"Ileal epithelial cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[470, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot2_ileal.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_ileal.index.get_loc(gene_name)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## MYELOID CELLS\n",
    "\n",
    "# # Cells charts\n",
    "# # creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "# stacked_row_chart_myel = StackedRowChart(\n",
    "#     title=\"Abundance plot\",\n",
    "#     params=[\"Patient\", \"Cell state\"],\n",
    "#     position=[930, 10],\n",
    "#     size=[580, 450]\n",
    "# )\n",
    "# # configuring the row chart\n",
    "# stacked_row_chart_myel.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "# #stacked_row_chart_myel.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "# #creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "# gene_name = \"HSPE1\"\n",
    "\n",
    "# dot_plot_myel = DotPlot(\n",
    "#     title= f\"Percent of gene expression per cell state\",\n",
    "#     #params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_myel.index.get_loc(gene_name)}\"],\n",
    "#     params = [\"Cell state\",{\"linkedDsName\":\"Myeloid genes\",\"maxItems\":10,\"type\":\"RowsAsColsQuery\"}],\n",
    "#     size=[450, 300],\n",
    "#     position=[10, 470]\n",
    "# )\n",
    "\n",
    "# # configuring the dot plot\n",
    "# dot_plot_myel.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_myel.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_myel.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_myel.set_color_scale(log_scale=False)\n",
    "# dot_plot_myel.set_color_legend(True, [40, 10])\n",
    "# dot_plot_myel.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "# box_plot_myel = BoxPlot(\n",
    "#     title=f\"Gene expression per cell state\",\n",
    "#     #params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_myel.index.get_loc(gene_name)}\"],\n",
    "#     params = [\"Cell state\", {\"linkedDsName\":\"Myeloid genes\",\"maxItems\":1,\"type\":\"RowsAsColsQuery\"}],\n",
    "#     size=[650, 300],\n",
    "#     position=[470, 470]\n",
    "# )\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot_myel = ScatterPlot(\n",
    "#     title=\"Myeloid cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[10, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot_myel.set_color_by(\"Cell state\")\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot2_myel = ScatterPlot(\n",
    "#     title=\"Myeloid cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[470, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot2_myel.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_myel.index.get_loc(gene_name)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## PLASMA CELLS\n",
    "\n",
    "# # Cells charts\n",
    "# # creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "# stacked_row_chart_plasma = StackedRowChart(\n",
    "#     title=\"Abundance plot\",\n",
    "#     params=[\"Patient\", \"Cell state\"],\n",
    "#     position=[930, 10],\n",
    "#     size=[580, 450]\n",
    "# )\n",
    "# # configuring the row chart\n",
    "# stacked_row_chart_plasma.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "# #stacked_row_chart_plasma.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "# #creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "# gene_name = \"HSPE1\"\n",
    "\n",
    "# dot_plot_plasma = DotPlot(\n",
    "#     title= f\"Percent of gene expression per cell state\",\n",
    "#     #params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_plasma.index.get_loc(gene_name)}\"],\n",
    "#     params = [\"Cell state\",{\"linkedDsName\":\"Plasma genes\",\"maxItems\":10,\"type\":\"RowsAsColsQuery\"}],\n",
    "#     size=[450, 300],\n",
    "#     position=[10, 470]\n",
    "# )\n",
    "\n",
    "# # configuring the dot plot\n",
    "# dot_plot_plasma.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_plasma.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_plasma.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_plasma.set_color_scale(log_scale=False)\n",
    "# dot_plot_plasma.set_color_legend(True, [40, 10])\n",
    "# dot_plot_plasma.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "# box_plot_plasma = BoxPlot(\n",
    "#     title=f\"Gene expression per cell state\",\n",
    "#     #params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_plasma.index.get_loc(gene_name)}\"],\n",
    "#     params = [\"Cell state\", {\"linkedDsName\":\"Plasma genes\",\"maxItems\":1,\"type\":\"RowsAsColsQuery\"}],\n",
    "#     size=[650, 300],\n",
    "#     position=[470, 470]\n",
    "# )\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot_plasma = ScatterPlot(\n",
    "#     title=\"Plasma cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[10, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot_plasma.set_color_by(\"Cell state\")\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot2_plasma = ScatterPlot(\n",
    "#     title=\"Plasma cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[470, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot2_plasma.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_plasma.index.get_loc(gene_name)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## INNATE LYMPHOID CELLS\n",
    "\n",
    "# # Cells charts\n",
    "# # creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "# stacked_row_chart_ilc = StackedRowChart(\n",
    "#     title=\"Abundance plot\",\n",
    "#     params=[\"Patient\", \"Cell state\"],\n",
    "#     position=[930, 10],\n",
    "#     size=[580, 450]\n",
    "# )\n",
    "# # configuring the row chart\n",
    "# stacked_row_chart_ilc.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "# #stacked_row_chart_ilc.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "# #creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "# gene_name = \"HSPE1\"\n",
    "\n",
    "# dot_plot_ilc = DotPlot(\n",
    "#     title= f\"Percent of gene expression per cell state\",\n",
    "#     #params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_ilc.index.get_loc(gene_name)}\"],\n",
    "#     params = [\"Cell state\",{\"linkedDsName\":\"Innate lymphoid genes\",\"maxItems\":10,\"type\":\"RowsAsColsQuery\"}],\n",
    "#     size=[450, 300],\n",
    "#     position=[10, 470]\n",
    "# )\n",
    "\n",
    "# # configuring the dot plot\n",
    "# dot_plot_ilc.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_ilc.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_ilc.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "# dot_plot_ilc.set_color_scale(log_scale=False)\n",
    "# dot_plot_ilc.set_color_legend(True, [40, 10])\n",
    "# dot_plot_ilc.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "# box_plot_ilc = BoxPlot(\n",
    "#     title=f\"Gene expression per cell state\",\n",
    "#     params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_ilc.index.get_loc(gene_name)}\"],\n",
    "#     size=[650, 300],\n",
    "#     position=[470, 470]\n",
    "# )\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot_ilc = ScatterPlot(\n",
    "#     title=\"Innate lymphoid cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[10, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot_ilc.set_color_by(\"Cell state\")\n",
    "\n",
    "# # creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "# scatter_plot2_ilc = ScatterPlot(\n",
    "#     title=\"Innate lymphoid cells\",\n",
    "#     params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "#     size=[450, 450],\n",
    "#     position=[470, 10],\n",
    "#     default_color=\"#377eb8\",\n",
    "#     brush=\"default\",\n",
    "#     on_filter=\"hide\",\n",
    "#     radius=5,\n",
    "#     opacity=0.8,\n",
    "# )\n",
    "\n",
    "# # configuring the scatter plot\n",
    "# scatter_plot2_ilc.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_ilc.index.get_loc(gene_name)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# setting up and serving the MDV project\n",
    "base = os.path.expanduser('~/mdv')\n",
    "project_path = os.path.join(base, 'taurus_new_test') # defining the location where the project metadata will be stored\n",
    "p = MDVProject(os.path.expanduser(project_path), delete_existing=True)\n",
    "\n",
    "# # adding the two data sources to the project\n",
    "# # B cells\n",
    "p.add_datasource(\"B cells\", cells_df_bcell)\n",
    "p.add_datasource(\"B genes\", gene_table_bcell)\n",
    "\n",
    "# # # CD4T cells\n",
    "# p.add_datasource(\"CD4T cells\", cells_df_cd4t)\n",
    "# p.add_datasource(\"CD4T genes\", gene_table_cd4t)\n",
    "\n",
    "# # # CD8T cells\n",
    "# p.add_datasource(\"CD8T cells\", cells_df_cd8t)\n",
    "# p.add_datasource(\"CD8T genes\", gene_table_cd8t)\n",
    "\n",
    "# # Colonic epithelial cells\n",
    "# p.add_datasource(\"Colonic epithelial cells\", cells_df_col_epi)\n",
    "# p.add_datasource(\"Colonic epithelial genes\", gene_table_col_epi)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # # Endothelial cells\n",
    "# p.add_datasource(\"Endothelial cells\", cells_df_endo)\n",
    "# p.add_datasource(\"Endothelial genes\", gene_table_endo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Fibroblast cells\n",
    "# p.add_datasource(\"Fibroblast pericyte cells\", cells_df_fib)\n",
    "# p.add_datasource(\"Fibroblast pericyte genes\", gene_table_fib)\n",
    "\n",
    "# # Ileal epithelial cells\n",
    "# p.add_datasource(\"Ileal epithelial cells\", cells_df_ileal)\n",
    "# p.add_datasource(\"Ileal epithelial genes\", gene_table_ileal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Innate lymphoid cells\n",
    "# p.add_datasource(\"Innate lymphoid cells\", cells_df_ilc)\n",
    "# p.add_datasource(\"Innate lymphoid genes\", gene_table_ilc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Myeloid cells\n",
    "# p.add_datasource(\"Myeloid cells\", cells_df_myel)\n",
    "# p.add_datasource(\"Myeloid genes\", gene_table_myel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Plasma cells\n",
    "# p.add_datasource(\"Plasma cells\", cells_df_plasma)\n",
    "# p.add_datasource(\"Plasma genes\", gene_table_plasma)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# creating the link between the two datasets so that selecting a subset of genes to add the expression in cells is enabled\n",
    "p.add_rows_as_columns_link(\"B cells\",\"B genes\",\"gene_id\",\"Gene expression\")\n",
    "p.add_rows_as_columns_subgroup(\"B cells\",\"B genes\",\"Gene expression\",bcell_data.layers[\"counts\"]) #add the gene expression "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# p.add_rows_as_columns_link(\"CD4T cells\",\"CD4T genes\",\"gene_id\",\"Gene expression\")\n",
    "# p.add_rows_as_columns_subgroup(\"CD4T cells\",\"CD4T genes\",\"Gene expression\",cd4tcell_data.layers[\"counts\"]) #add the gene expression "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# p.add_rows_as_columns_link(\"CD8T cells\",\"CD8T genes\",\"gene_id\",\"Gene expression\")\n",
    "# p.add_rows_as_columns_subgroup(\"CD8T cells\",\"CD8T genes\",\"Gene expression\",cd8tcell_data.layers[\"counts\"]) #add the gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# p.add_rows_as_columns_link(\"Endothelial cells\",\"Endothelial genes\",\"gene_id\",\"Gene expression\")\n",
    "# p.add_rows_as_columns_subgroup(\"Endothelial cells\",\"Endothelial genes\",\"Gene expression\",endothelial_data.layers[\"counts\"]) #add the gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# p.add_rows_as_columns_link(\"Fibroblast pericyte cells\",\"Fibroblast pericyte genes\",\"gene_id\",\"Gene expression\")\n",
    "# p.add_rows_as_columns_subgroup(\"Fibroblast pericyte cells\",\"Fibroblast pericyte genes\",\"Gene expression\",fib_peri_data.layers[\"counts\"]) #add the gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# p.add_rows_as_columns_link(\"Ileal epithelial cells\",\"Ileal epithelial genes\",\"gene_id\",\"Gene expression\")\n",
    "# p.add_rows_as_columns_subgroup(\"Ileal epithelial cells\",\"Ileal epithelial genes\",\"Gene expression\",epi_ileal_data.layers[\"counts\"]) #add the gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# p.add_rows_as_columns_link(\"Innate lymphoid cells\",\"Innate lymphoid genes\",\"gene_id\",\"Gene expression\")\n",
    "# p.add_rows_as_columns_subgroup(\"Innate lymphoid cells\",\"Innate lymphoid genes\",\"Gene expression\",ilc_data.layers[\"counts\"]) #add the gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# p.add_rows_as_columns_link(\"Myeloid cells\",\"Myeloid genes\",\"gene_id\",\"Gene expression\")\n",
    "# p.add_rows_as_columns_subgroup(\"Myeloid cells\",\"Myeloid genes\",\"Gene expression\",myeloid_data.layers[\"counts\"]) #add the gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# p.add_rows_as_columns_link(\"Plasma cells\",\"Plasma genes\",\"gene_id\",\"Gene expression\")\n",
    "# p.add_rows_as_columns_subgroup(\"Plasma cells\",\"Plasma genes\",\"Gene expression\",plasmacell_data.layers[\"counts\"]) #add the gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# p.add_rows_as_columns_link(\"Colonic epithelial cells\",\"Colonic epithelial genes\",\"gene_id\",\"Gene expression\")\n",
    "# p.add_rows_as_columns_subgroup(\"Colonic epithelial cells\",\"Colonic epithelial genes\",\"Gene expression\",epithelium_colonic_data.layers[\"counts\"]) #add the gene expression\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # converting the chart implementation outputs to JSON and setting up the project view\n",
    "list_charts_cells_bcell = []\n",
    "list_charts_cells_cd4t = []\n",
    "list_charts_cells_cd8t = []\n",
    "list_charts_cells_col_epi = []\n",
    "list_charts_cells_endo = []\n",
    "list_charts_cells_fib = []\n",
    "list_charts_cells_ileal = []\n",
    "list_charts_cells_ilc = []\n",
    "list_charts_cells_myel = []\n",
    "list_charts_cells_plasma = []\n",
    "\n",
    "# cells panel\n",
    "list_charts_cells_bcell.extend([stacked_row_chart_bcell.plot_data, dot_plot_bcell.plot_data, box_plot_bcell.plot_data, \n",
    "                                scatter_plot_bcell.plot_data, scatter_plot2_bcell.plot_data])\n",
    "# list_charts_cells_cd4t.extend([stacked_row_chart_cd4t.plot_data, dot_plot_cd4t.plot_data, box_plot_cd4t.plot_data, \n",
    "#                                 scatter_plot_cd4t.plot_data, scatter_plot2_cd4t.plot_data])\n",
    "# list_charts_cells_cd8t.extend([stacked_row_chart_cd8t.plot_data, dot_plot_cd8t.plot_data, box_plot_cd8t.plot_data, \n",
    "#                                 scatter_plot_cd8t.plot_data, scatter_plot2_cd8t.plot_data])\n",
    "# # list_charts_cells_col_epi.extend([stacked_row_chart_col_epi.plot_data, dot_plot_col_epi.plot_data, box_plot_col_epi.plot_data, \n",
    "# #                                 scatter_plot_col_epi.plot_data, scatter_plot2_col_epi.plot_data])\n",
    "# list_charts_cells_endo.extend([stacked_row_chart_endo.plot_data, dot_plot_endo.plot_data, box_plot_endo.plot_data, \n",
    "#                                 scatter_plot_endo.plot_data, scatter_plot2_endo.plot_data])\n",
    "# list_charts_cells_fib.extend([stacked_row_chart_fib.plot_data, dot_plot_fib.plot_data, box_plot_fib.plot_data, \n",
    "#                                 scatter_plot_fib.plot_data, scatter_plot2_fib.plot_data])\n",
    "# list_charts_cells_ileal.extend([stacked_row_chart_ileal.plot_data, dot_plot_ileal.plot_data, box_plot_ileal.plot_data, \n",
    "#                                 scatter_plot_ileal.plot_data, scatter_plot2_ileal.plot_data])\n",
    "# list_charts_cells_ilc.extend([stacked_row_chart_ilc.plot_data, dot_plot_ilc.plot_data, box_plot_ilc.plot_data, \n",
    "#                                 scatter_plot_ilc.plot_data, scatter_plot2_ilc.plot_data])\n",
    "# list_charts_cells_myel.extend([stacked_row_chart_myel.plot_data, dot_plot_myel.plot_data, box_plot_myel.plot_data, \n",
    "#                                 scatter_plot_myel.plot_data, scatter_plot2_myel.plot_data])\n",
    "# list_charts_cells_plasma.extend([stacked_row_chart_plasma.plot_data, dot_plot_plasma.plot_data, box_plot_plasma.plot_data, \n",
    "#                                 scatter_plot_plasma.plot_data, scatter_plot2_plasma.plot_data])\n",
    "\n",
    "# setting the config combining the two panels\n",
    "view_config_bcell = {'initialCharts': {\"B cells\": list_charts_cells_bcell}}#, \"B genes\": []}}\n",
    "# view_config_cd4t = {'initialCharts': {\"CD4T cells\": list_charts_cells_cd4t, \"CD4T genes\": []}}\n",
    "# view_config_cd8t = {'initialCharts': {\"CD8T cells\": list_charts_cells_cd8t, \"CD8T genes\": []}}\n",
    "# #view_config_col_epi = {'initialCharts': {\"Colonic epithelial cells\": list_charts_cells_col_epi, \"Colonic epithelial genes\": []}}\n",
    "# view_config_endo = {'initialCharts': {\"Endothelial cells\": list_charts_cells_endo, \"Endothelial genes\": []}}\n",
    "# view_config_fib = {'initialCharts': {\"Fibroblast pericyte cells\": list_charts_cells_fib, \"Fibroblast pericyte genes\": []}}\n",
    "# view_config_ileal = {'initialCharts': {\"Ileal epithelial cells\": list_charts_cells_ileal, \"Ileal epithelial genes\": []}}\n",
    "# view_config_ilc = {'initialCharts': {\"Innate lymphoid cells\": list_charts_cells_ilc, \"Innate lymphoid genes\": []}}\n",
    "# view_config_myel = {'initialCharts': {\"Myeloid cells\": list_charts_cells_myel, \"Myeloid genes\": []}}\n",
    "# view_config_plasma = {'initialCharts': {\"Plasma cells\": list_charts_cells_plasma, \"Plasma genes\": []}}\n",
    "\n",
    "# adding the view to the project configuration\n",
    "p.set_view(\"B cells\", view_config_bcell)\n",
    "# p.set_view(\"CD4T cells\", view_config_cd4t)\n",
    "# p.set_view(\"CD8T cells\", view_config_cd8t)\n",
    "# #p.set_view(\"Colonic epithelial cells\", view_config_col_epi)\n",
    "# p.set_view(\"Endothelial cells\", view_config_endo)\n",
    "# p.set_view(\"Fibroblast pericyte cells\", view_config_fib)\n",
    "# p.set_view(\"Ileal epithelial cells\", view_config_ileal)\n",
    "# p.set_view(\"Innate lymphoid cells\", view_config_ilc)\n",
    "# p.set_view(\"Myeloid cells\", view_config_myel)\n",
    "# p.set_view(\"Plasma cells\", view_config_plasma)\n",
    "\n",
    "p.set_editable(True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# setting keys and variables\n",
      "# Crawl the local repository to get a list of relevant file paths\n",
      "Block 'b1: Local repo crawling' took 0.0152 seconds\n",
      "# Initialize a text splitter for chunking the code strings\n",
      "# Split the code documents into chunks using the text splitter\n",
      "Block 'b2: Text splitter initialising' took 0.0003 seconds\n",
      "# Initialize an instance of the OpenAIEmbeddings class\n",
      "Block 'b3: Embeddings creating' took 0.0442 seconds\n",
      "client=<openai.resources.embeddings.Embeddings object at 0x381e8c680> async_client=<openai.resources.embeddings.AsyncEmbeddings object at 0x381e8e570> model='text-embedding-3-large' dimensions=None deployment='text-embedding-ada-002' openai_api_version=None openai_api_base=None openai_api_type=None openai_proxy=None embedding_ctx_length=8191 openai_api_key=SecretStr('**********') openai_organization=None allowed_special=None disallowed_special=None chunk_size=1000 max_retries=2 request_timeout=None headers=None tiktoken_enabled=True tiktoken_model_name=None show_progress_bar=False model_kwargs={} skip_empty=False default_headers=None default_query=None retry_min_seconds=4 retry_max_seconds=20 http_client=None http_async_client=None check_embedding_ctx_length=True\n",
      "# Create an index from the embedded code chunks\n",
      "# Use FAISS (Facebook AI Similarity Search) to create a searchable index\n",
      "Block 'b4: FAISS database creating' took 1.7032 seconds\n",
      "# Initialize the retriever from the FAISS index\n",
      "Block 'b5: Retriever creating' took 0.0000 seconds\n",
      "created Flask <Flask 'mdvtools.server'>\n",
      "socketio initialized\n",
      "websocket/chat enabled for \n",
      " * Serving Flask app 'mdvtools.server'\n",
      " * Debug mode: on\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.\n",
      " * Running on all addresses (0.0.0.0)\n",
      " * Running on http://127.0.0.1:5050\n",
      " * Running on http://129.67.46.69:5050\n",
      "Press CTRL+C to quit\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:05] \"GET / HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:05] \"GET /static/assets/mdv.css HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:05] \"GET /static/js/mdv.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:05] \"GET /static/assets/bot-message-square-DdYfrx9z.js HTTP/1.1\" 304 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recieved request to project_index\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [27/Nov/2024 16:14:05] \"GET /static/assets/csvWorker-e8cMUEqV.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:05] \"GET /datasources.json HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:05] \"GET /state.json HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:05] \"GET /views.json HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:05] \"POST /get_view HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:05] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:05] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:05] \"GET /static/img/fa-solid-900.woff2 HTTP/1.1\" 304 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing state response. websocket: True\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [27/Nov/2024 16:14:06] \"GET /chat_log.json HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:08] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:10] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:12] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:14] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:16] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:18] \"POST /get_view HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:18] \"POST /get_data HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:18] \"POST /get_data HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:18] \"POST /get_data HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:18] \"GET /static/assets/catWorker-D5lhlpxE.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:18] \"GET /static/assets/binWorker-DIC2QqPk.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:18] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:18] \"GET /static/assets/catColWorker-BRK20MZB.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:18] \"GET /static/assets/binWorker-DIC2QqPk.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:18] \"GET /static/assets/boxPlotWorker-BvOiGWqU.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:18] \"GET /static/assets/binWorker-DIC2QqPk.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:20] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:22] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:23] \"GET /static/img/roboto-latin-400-normal.woff2 HTTP/1.1\" 304 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chat_init\n",
      "[llm taurus_new_test] The project has more than one datasource, only the first one will be used\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/langchain_experimental/agents/agent_toolkits/pandas/base.py:283: UserWarning: Received additional kwargs {'handle_parse_errors': True} which are no longer supported.\n",
      "  warnings.warn(\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:23] \"POST /chat_init HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Block 'b6: Initialising LLM for RAG' took 0.0287 seconds\n",
      "Block 'b7: Initialising LLM for agent' took 0.0276 seconds\n",
      "Block 'b8: Pandas agent creating' took 0.0240 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [27/Nov/2024 16:14:24] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:26] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:28] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:30] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:32] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:34] \"GET /chat_log.json HTTP/1.1\" 304 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[llm taurus_new_test] Asking the LLM: 'Can you please generate a scatter plot for the PCA components?'\n",
      "\n",
      "Based on the question asked and the dataframes provided, please perform the following steps:\n",
      "\n",
      "1. Identify the data asked for in the question.\n",
      "2. Based on step 1, find the relevant column names in the dataframes provided based on the information identified earlier in the question asked regarding data.\n",
      "3. The relevant column names can be a combination from the two dataframes.\n",
      "4. Provide the relevant column names from step 2 in a list.\n",
      "5. If the question asks for a gene name, do provide a gene name in addition to the relevant columns names.\n",
      "\n",
      "Question: Can you please generate a scatter plot for the PCA components?\n",
      "Block 'b9a: Pandas agent invoking' took 0.0011 seconds\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new AgentExecutor chain...\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [27/Nov/2024 16:14:36] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:38] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:40] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:42] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:44] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:46] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:48] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:50] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:52] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:54] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:14:56] \"GET /chat_log.json HTTP/1.1\" 304 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32;1m\u001b[1;3mTo generate a scatter plot for the PCA components, we need to identify the relevant columns in the dataframes that correspond to PCA components. However, based on the provided dataframes, there are no explicit columns labeled as PCA components. Typically, PCA components would be derived from a dimensionality reduction process applied to the data, and they would not be directly present in the initial dataframes.\n",
      "\n",
      "Given this, we need to perform PCA on the data to obtain the PCA components. We can use the numerical columns from `df1` for this purpose, as `df2` seems to contain gene expression data which is not directly relevant for PCA unless we are specifically analyzing gene expression patterns.\n",
      "\n",
      "Let's proceed with the following steps:\n",
      "1. Select numerical columns from `df1` that could be used for PCA.\n",
      "2. Perform PCA on the selected columns.\n",
      "3. Generate a scatter plot for the first two PCA components.\n",
      "\n",
      "Action: Perform PCA on the numerical columns of `df1` and generate a scatter plot for the first two PCA components.\n",
      "Action Input: \n",
      "```python\n",
      "from sklearn.decomposition import PCA\n",
      "import matplotlib.pyplot as plt\n",
      "\n",
      "# Select numerical columns from df1 for PCA\n",
      "numerical_cols = df1.select_dtypes(include=['float64', 'int64']).columns\n",
      "\n",
      "# Perform PCA\n",
      "pca = PCA(n_components=2)\n",
      "pca_components = pca.fit_transform(df1[numerical_cols])\n",
      "\n",
      "# Generate scatter plot\n",
      "plt.figure(figsize=(8, 6))\n",
      "plt.scatter(pca_components[:, 0], pca_components[:, 1], alpha=0.5)\n",
      "plt.title('PCA Scatter Plot')\n",
      "plt.xlabel('PCA Component 1')\n",
      "plt.ylabel('PCA Component 2')\n",
      "plt.show()\n",
      "```\u001b[0mPerform PCA on the numerical columns of `df1` and generate a scatter plot for the first two PCA components. is not a valid tool, try one of [python_repl_ast]."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [27/Nov/2024 16:14:58] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:15:00] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:15:02] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:15:04] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:15:07] \"GET /chat_log.json HTTP/1.1\" 304 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32;1m\u001b[1;3mTo generate a scatter plot for the PCA components, I need to perform PCA on the numerical columns of `df1` and then plot the first two components. Let's proceed with this step-by-step process.\n",
      "\n",
      "Action: Perform PCA on the numerical columns of `df1` and generate a scatter plot for the first two PCA components.\n",
      "Action Input: \n",
      "```python\n",
      "from sklearn.decomposition import PCA\n",
      "import matplotlib.pyplot as plt\n",
      "\n",
      "# Select numerical columns from df1 for PCA\n",
      "numerical_cols = df1.select_dtypes(include=['float64', 'int64']).columns\n",
      "\n",
      "# Perform PCA\n",
      "pca = PCA(n_components=2)\n",
      "pca_components = pca.fit_transform(df1[numerical_cols])\n",
      "\n",
      "# Generate scatter plot\n",
      "plt.figure(figsize=(8, 6))\n",
      "plt.scatter(pca_components[:, 0], pca_components[:, 1], alpha=0.5)\n",
      "plt.title('PCA Scatter Plot')\n",
      "plt.xlabel('PCA Component 1')\n",
      "plt.ylabel('PCA Component 2')\n",
      "plt.show()\n",
      "```\u001b[0mPerform PCA on the numerical columns of `df1` and generate a scatter plot for the first two PCA components. is not a valid tool, try one of [python_repl_ast].\u001b[32;1m\u001b[1;3mTo generate a scatter plot for the PCA components, I need to perform PCA on the numerical columns of `df1` and then plot the first two components. Let's proceed with this step-by-step process using the `python_repl_ast` tool.\n",
      "\n",
      "Action: Perform PCA on the numerical columns of `df1` and generate a scatter plot for the first two PCA components.\n",
      "Action Input: \n",
      "```python\n",
      "from sklearn.decomposition import PCA\n",
      "import matplotlib.pyplot as plt\n",
      "import pandas as pd\n",
      "\n",
      "# Sample data for df1\n",
      "data = {\n",
      "    'Sample ID': ['CID003352-2'] * 5,\n",
      "    'doublet_scores': [0.0297433, 0.0629283, 0.0385867, 0.0396102, 0.0275013],\n",
      "    'n_genes_by_counts': [1169, 841, 1560, 1394, 1189],\n",
      "    'total_counts': [3650, 2126, 4770, 4081, 3290],\n",
      "    'total_counts_mt': [556, 331, 623, 477, 463],\n",
      "    'pct_counts_mt': [15.2329, 15.5691, 13.0608, 11.6883, 14.0729],\n",
      "    'total_counts_rp': [941, 293, 932, 1068, 656],\n",
      "    'pct_counts_rp': [25.7808, 13.7817, 19.5388, 26.1701, 19.9392],\n",
      "    'S_score': [-0.0814609, -0.034859, -0.145446, -0.221775, -0.158715],\n",
      "    'G2M_score': [-0.170746, -0.129371, -0.270396, -0.277972, -0.183566],\n",
      "    'UMAP 1': [0.135776, 3.20875, 1.16571, 1.8318, -0.556194],\n",
      "    'UMAP 2': [-1.97701, -1.22797, -0.499263, -1.06328, -2.28645]\n",
      "}\n",
      "\n",
      "df1 = pd.DataFrame(data)\n",
      "\n",
      "# Select numerical columns from df1 for PCA\n",
      "numerical_cols = df1.select_dtypes(include=['float64', 'int64']).columns\n",
      "\n",
      "# Perform PCA\n",
      "pca = PCA(n_components=2)\n",
      "pca_components = pca.fit_transform(df1[numerical_cols])\n",
      "\n",
      "# Generate scatter plot\n",
      "plt.figure(figsize=(8, 6))\n",
      "plt.scatter(pca_components[:, 0], pca_components[:, 1], alpha=0.5)\n",
      "plt.title('PCA Scatter Plot')\n",
      "plt.xlabel('PCA Component 1')\n",
      "plt.ylabel('PCA Component 2')\n",
      "plt.show()\n",
      "```\u001b[0mPerform PCA on the numerical columns of `df1` and generate a scatter plot for the first two PCA components. is not a valid tool, try one of [python_repl_ast]."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [27/Nov/2024 16:16:07] \"GET /chat_log.json HTTP/1.1\" 304 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32;1m\u001b[1;3mTo generate a scatter plot for the PCA components, I need to perform PCA on the numerical columns of `df1` and then plot the first two components. Let's proceed with this step-by-step process using the `python_repl_ast` tool.\n",
      "\n",
      "Action: Perform PCA on the numerical columns of `df1` and generate a scatter plot for the first two PCA components.\n",
      "Action Input: \n",
      "```python\n",
      "from sklearn.decomposition import PCA\n",
      "import matplotlib.pyplot as plt\n",
      "import pandas as pd\n",
      "\n",
      "# Sample data for df1\n",
      "data = {\n",
      "    'Sample ID': ['CID003352-2'] * 5,\n",
      "    'doublet_scores': [0.0297433, 0.0629283, 0.0385867, 0.0396102, 0.0275013],\n",
      "    'n_genes_by_counts': [1169, 841, 1560, 1394, 1189],\n",
      "    'total_counts': [3650, 2126, 4770, 4081, 3290],\n",
      "    'total_counts_mt': [556, 331, 623, 477, 463],\n",
      "    'pct_counts_mt': [15.2329, 15.5691, 13.0608, 11.6883, 14.0729],\n",
      "    'total_counts_rp': [941, 293, 932, 1068, 656],\n",
      "    'pct_counts_rp': [25.7808, 13.7817, 19.5388, 26.1701, 19.9392],\n",
      "    'S_score': [-0.0814609, -0.034859, -0.145446, -0.221775, -0.158715],\n",
      "    'G2M_score': [-0.170746, -0.129371, -0.270396, -0.277972, -0.183566],\n",
      "    'UMAP 1': [0.135776, 3.20875, 1.16571, 1.8318, -0.556194],\n",
      "    'UMAP 2': [-1.97701, -1.22797, -0.499263, -1.06328, -2.28645]\n",
      "}\n",
      "\n",
      "df1 = pd.DataFrame(data)\n",
      "\n",
      "# Select numerical columns from df1 for PCA\n",
      "numerical_cols = df1.select_dtypes(include=['float64', 'int64']).columns\n",
      "\n",
      "# Perform PCA\n",
      "pca = PCA(n_components=2)\n",
      "pca_components = pca.fit_transform(df1[numerical_cols])\n",
      "\n",
      "# Generate scatter plot\n",
      "plt.figure(figsize=(8, 6))\n",
      "plt.scatter(pca_components[:, 0], pca_components[:, 1], alpha=0.5)\n",
      "plt.title('PCA Scatter Plot')\n",
      "plt.xlabel('PCA Component 1')\n",
      "plt.ylabel('PCA Component 2')\n",
      "plt.show()\n",
      "```\u001b[0mPerform PCA on the numerical columns of `df1` and generate a scatter plot for the first two PCA components. is not a valid tool, try one of [python_repl_ast].\u001b[32;1m\u001b[1;3mTo generate a scatter plot for the PCA components, I need to perform PCA on the numerical columns of `df1` and then plot the first two components. Let's proceed with this step-by-step process using the `python_repl_ast` tool.\n",
      "\n",
      "Action: Perform PCA on the numerical columns of `df1` and generate a scatter plot for the first two PCA components.\n",
      "Action Input: \n",
      "```python\n",
      "from sklearn.decomposition import PCA\n",
      "import matplotlib.pyplot as plt\n",
      "import pandas as pd\n",
      "\n",
      "# Sample data for df1\n",
      "data = {\n",
      "    'Sample ID': ['CID003352-2'] * 5,\n",
      "    'doublet_scores': [0.0297433, 0.0629283, 0.0385867, 0.0396102, 0.0275013],\n",
      "    'n_genes_by_counts': [1169, 841, 1560, 1394, 1189],\n",
      "    'total_counts': [3650, 2126, 4770, 4081, 3290],\n",
      "    'total_counts_mt': [556, 331, 623, 477, 463],\n",
      "    'pct_counts_mt': [15.2329, 15.5691, 13.0608, 11.6883, 14.0729],\n",
      "    'total_counts_rp': [941, 293, 932, 1068, 656],\n",
      "    'pct_counts_rp': [25.7808, 13.7817, 19.5388, 26.1701, 19.9392],\n",
      "    'S_score': [-0.0814609, -0.034859, -0.145446, -0.221775, -0.158715],\n",
      "    'G2M_score': [-0.170746, -0.129371, -0.270396, -0.277972, -0.183566],\n",
      "    'UMAP 1': [0.135776, 3.20875, 1.16571, 1.8318, -0.556194],\n",
      "    'UMAP 2': [-1.97701, -1.22797, -0.499263, -1.06328, -2.28645]\n",
      "}\n",
      "\n",
      "df1 = pd.DataFrame(data)\n",
      "\n",
      "# Select numerical columns from df1 for PCA\n",
      "numerical_cols = df1.select_dtypes(include=['float64', 'int64']).columns\n",
      "\n",
      "# Perform PCA\n",
      "pca = PCA(n_components=2)\n",
      "pca_components = pca.fit_transform(df1[numerical_cols])\n",
      "\n",
      "# Generate scatter plot\n",
      "plt.figure(figsize=(8, 6))\n",
      "plt.scatter(pca_components[:, 0], pca_components[:, 1], alpha=0.5)\n",
      "plt.title('PCA Scatter Plot')\n",
      "plt.xlabel('PCA Component 1')\n",
      "plt.ylabel('PCA Component 2')\n",
      "plt.show()\n",
      "```\u001b[0mPerform PCA on the numerical columns of `df1` and generate a scatter plot for the first two PCA components. is not a valid tool, try one of [python_repl_ast]."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [27/Nov/2024 16:16:41] \"POST /chat HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"GET /?view=B+cells HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"GET /static/assets/mdv.css HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"GET /static/js/mdv.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"GET /static/assets/bot-message-square-DdYfrx9z.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"GET /static/assets/csvWorker-e8cMUEqV.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"GET /datasources.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"GET /state.json HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"GET /views.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"POST /get_view HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"GET /static/img/fa-solid-900.woff2 HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"POST /get_data HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"POST /get_data HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"POST /get_data HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recieved request to project_index\n",
      "processing state response. websocket: True\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"GET /static/assets/catWorker-D5lhlpxE.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"GET /static/assets/binWorker-DIC2QqPk.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:46] \"GET /static/assets/catColWorker-BRK20MZB.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:47] \"GET /static/assets/binWorker-DIC2QqPk.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:47] \"GET /static/assets/boxPlotWorker-BvOiGWqU.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:16:47] \"GET /static/assets/binWorker-DIC2QqPk.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 16:24:54] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 17:03:31] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [27/Nov/2024 17:21:04] \"GET /chat_log.json HTTP/1.1\" 304 -\n"
     ]
    }
   ],
   "source": [
    "# serving the project\n",
    "p.serve()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#p.convert_to_static_page(\"../../../../../Desktop/TAURUS_static_new_3/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venvjupyter",
   "language": "python",
   "name": "venvjupyter"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
