{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Processing, Clustering and Visualising in MDV 3k PBMCs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The part of this tutorial dedicated to analysing the data is based on the scanpy tutorial \"Preprocessing and clustering 3k PBMCs (legacy workflow)\" found here: https://scanpy.readthedocs.io/en/stable/tutorials/index.html . If you are interested in understanding the data analysis part of this tutorial more, please check the scanpy tutorial. The data consists of 3k PBMCs from a Healthy Donor and are available from 10x Genomics. On a unix system, you can uncomment and run the following to download and unpack the data. Alternatively you can download the data from http://cf.10xgenomics.com/samples/cell-exp/1.1.0/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz, create a new directory named \"data\" in the same level that this jupyter notebook is saved and run from, and unzip the data within the \"data\" directory."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Importing the required packages for data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Importing the required packages for MDV set up and visualisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from mdvtools.mdvproject import MDVProject\n",
    "from mdvtools.charts.dot_plot import DotPlot\n",
    "from mdvtools.charts.scatter_plot import ScatterPlot\n",
    "from mdvtools.charts.box_plot import BoxPlot\n",
    "from mdvtools.charts.stacked_row_plot import StackedRowChart"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data analysis section"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.10.1 anndata==0.9.2 umap==0.5.6 numpy==1.26.4 scipy==1.13.0 pandas==2.2.2 scikit-learn==1.4.2 statsmodels==0.14.2 igraph==0.11.4 pynndescent==0.5.12\n"
     ]
    }
   ],
   "source": [
    "# scanpy parameters for feedback level setting\n",
    "sc.settings.verbosity = 3  # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_header() # printing a header of introductory information about the environment/library used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reading the 3k 10x data\n",
    "\n",
    "# adata_bcell = sc.read_h5ad(\"../../../../../Documents/TAURUS_data/bcell_viz_ready_revised.h5ad\",)\n",
    "\n",
    "bcell_data = sc.read_h5ad('../../../../../Documents/TAURUS_data/bcell_viz_ready_revised.h5ad')\n",
    "cd4tcell_data = sc.read_h5ad('../../../../../Documents/TAURUS_data/cd4tcell_viz_ready_revised.h5ad')\n",
    "cd8tcell_data = sc.read_h5ad('../../../../../Documents/TAURUS_data/cd8tcell_viz_ready_revised.h5ad')\n",
    "epithelium_colonic_data = sc.read_h5ad('../../../../../Documents/TAURUS_data/epithelium_colonic_viz_ready_revised.h5ad')\n",
    "endothelial_data = sc.read_h5ad('../../../../../Documents/TAURUS_data/endothelium_glial_viz_ready_revised.h5ad')\n",
    "epi_ileal_data = sc.read_h5ad('../../../../../Documents/TAURUS_data/epi_ileal_viz_ready_revised.h5ad')\n",
    "fib_peri_data = sc.read_h5ad('../../../../../Documents/TAURUS_data/fib_peri_viz_ready_revised.h5ad')\n",
    "ilc_data = sc.read_h5ad('../../../../../Documents/TAURUS_data/ilc_viz_ready_revised.h5ad')\n",
    "myeloid_data = sc.read_h5ad('../../../../../Documents/TAURUS_data/myeloid_viz_ready_revised.h5ad')\n",
    "plasmacell_data = sc.read_h5ad('../../../../../Documents/TAURUS_data/plasmacell_viz_ready_revised.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Data preprocessing\n",
    "\n",
    "The lines sc.pp.normalise and sc.pplog1p can be commented out when the new final data are available"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "normalizing counts per cell\n",
      "    finished (0:00:02)\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n"
     ]
    }
   ],
   "source": [
    "# B cells\n",
    "# Saving count data\n",
    "bcell_data.layers[\"counts\"] = bcell_data.X.copy()\n",
    "# total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "sc.pp.normalize_total(bcell_data, target_sum=1e4)\n",
    "# logarithmising the data\n",
    "sc.pp.log1p(bcell_data)\n",
    "\n",
    "# CD4T cells\n",
    "# Saving count data\n",
    "cd4tcell_data.layers[\"counts\"] = cd4tcell_data.X.copy()\n",
    "# total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "sc.pp.normalize_total(cd4tcell_data, target_sum=1e4)\n",
    "# logarithmising the data\n",
    "sc.pp.log1p(cd4tcell_data)\n",
    "\n",
    "# CD8T cells\n",
    "# Saving count data\n",
    "cd8tcell_data.layers[\"counts\"] = cd8tcell_data.X.copy()\n",
    "# total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "sc.pp.normalize_total(cd8tcell_data, target_sum=1e4)\n",
    "# logarithmising the data\n",
    "sc.pp.log1p(cd8tcell_data)\n",
    "\n",
    "# Endothelium cells\n",
    "# Saving count data\n",
    "endothelial_data.layers[\"counts\"] = endothelial_data.X.copy()\n",
    "# total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "sc.pp.normalize_total(endothelial_data, target_sum=1e4)\n",
    "# logarithmising the data\n",
    "sc.pp.log1p(endothelial_data)\n",
    "\n",
    "# epithelial cells\n",
    "# Saving count data\n",
    "epi_ileal_data.layers[\"counts\"] = epi_ileal_data.X.copy()\n",
    "# total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "sc.pp.normalize_total(epi_ileal_data, target_sum=1e4)\n",
    "# logarithmising the data\n",
    "sc.pp.log1p(epi_ileal_data)\n",
    "\n",
    "# Epithelium colonic cells\n",
    "# Saving count data\n",
    "epithelium_colonic_data.layers[\"counts\"] = epithelium_colonic_data.X.copy()\n",
    "# total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "sc.pp.normalize_total(epithelium_colonic_data, target_sum=1e4)\n",
    "# logarithmising the data\n",
    "sc.pp.log1p(epithelium_colonic_data)\n",
    "\n",
    "# Fiber cells\n",
    "# Saving count data\n",
    "fib_peri_data.layers[\"counts\"] = fib_peri_data.X.copy()\n",
    "# total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "sc.pp.normalize_total(fib_peri_data, target_sum=1e4)\n",
    "# logarithmising the data\n",
    "sc.pp.log1p(fib_peri_data)\n",
    "\n",
    "# Ilc cells\n",
    "# Saving count data\n",
    "ilc_data.layers[\"counts\"] = ilc_data.X.copy()\n",
    "# total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "sc.pp.normalize_total(ilc_data, target_sum=1e4)\n",
    "# logarithmising the data\n",
    "sc.pp.log1p(ilc_data)\n",
    "\n",
    "# Myeloid cells\n",
    "# Saving count data\n",
    "myeloid_data.layers[\"counts\"] = myeloid_data.X.copy()\n",
    "# total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "sc.pp.normalize_total(myeloid_data, target_sum=1e4)\n",
    "# logarithmising the data\n",
    "sc.pp.log1p(myeloid_data)\n",
    "\n",
    "# Plasma cells\n",
    "# Saving count data\n",
    "plasmacell_data.layers[\"counts\"] = plasmacell_data.X.copy()\n",
    "# total-count normalising the anndata object to 10,000 reads per cell, so that counts become comparable among cells\n",
    "sc.pp.normalize_total(plasmacell_data, target_sum=1e4)\n",
    "# logarithmising the data\n",
    "sc.pp.log1p(plasmacell_data)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Converting the anndata object to pandas dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Â B CELLS\n",
    "# cells dataframe \n",
    "cells_df_bcell = pd.DataFrame(bcell_data.obs)\n",
    "\n",
    "# adding the umap data to the cells dataframe\n",
    "#umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "cells_df_bcell[\"UMAP 1\"] = np.array(bcell_data.obsm[\"X_umap\"])[:, 0]\n",
    "cells_df_bcell[\"UMAP 2\"] = np.array(bcell_data.obsm[\"X_umap\"])[:, 1]\n",
    "cells_df_bcell[\"Cell ID\"] = bcell_data.obs.index\n",
    "cells_df_bcell.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "                         \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "cells_df_bcell = cells_df_bcell.iloc[:, [0,16,17,18,19,20,21,22,23,24,25,26,27]]\n",
    "\n",
    "# genes dataframe\n",
    "gene_table_bcell = bcell_data.var\n",
    "gene_table_bcell[\"gene_id\"]=gene_table_bcell.index\n",
    "\n",
    "\n",
    "##Â CD4T CELLS\n",
    "# cells dataframe \n",
    "cells_df_cd4t = pd.DataFrame(cd4tcell_data.obs)\n",
    "\n",
    "# adding the umap data to the cells dataframe\n",
    "#umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "cells_df_cd4t[\"UMAP 1\"] = np.array(cd4tcell_data.obsm[\"X_umap\"])[:, 0]\n",
    "cells_df_cd4t[\"UMAP 2\"] = np.array(cd4tcell_data.obsm[\"X_umap\"])[:, 1]\n",
    "cells_df_cd4t[\"Cell ID\"] = cd4tcell_data.obs.index\n",
    "cells_df_cd4t.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "                         \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "cells_df_cd4t = cells_df_cd4t.iloc[:, [0,16,17,18,19,20,21,22,23,24,25,26,27]]\n",
    "\n",
    "# genes dataframe\n",
    "gene_table_cd4t = cd4tcell_data.var\n",
    "gene_table_cd4t[\"gene_id\"]=gene_table_cd4t.index\n",
    "\n",
    "##Â CD8T CELLS\n",
    "# cells dataframe \n",
    "cells_df_cd8t = pd.DataFrame(cd8tcell_data.obs)\n",
    "\n",
    "# adding the umap data to the cells dataframe\n",
    "#umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "cells_df_cd8t[\"UMAP 1\"] = np.array(cd8tcell_data.obsm[\"X_umap\"])[:, 0]\n",
    "cells_df_cd8t[\"UMAP 2\"] = np.array(cd8tcell_data.obsm[\"X_umap\"])[:, 1]\n",
    "cells_df_cd8t[\"Cell ID\"] = cd8tcell_data.obs.index\n",
    "cells_df_cd8t.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "                         \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "cells_df_cd8t = cells_df_cd8t.iloc[:, [0,16,17,18,19,20,21,22,23,24,25,26,27]]\n",
    "\n",
    "# genes dataframe\n",
    "gene_table_cd8t = cd8tcell_data.var\n",
    "gene_table_cd8t[\"gene_id\"]=gene_table_cd8t.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "## COLONIC EPITHELIAL CELLS\n",
    "# cells dataframe \n",
    "cells_df_col_epi = pd.DataFrame(epithelium_colonic_data.obs)\n",
    "\n",
    "# adding the umap data to the cells dataframe\n",
    "#umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "cells_df_col_epi[\"UMAP 1\"] = np.array(epithelium_colonic_data.obsm[\"X_umap\"])[:, 0]\n",
    "cells_df_col_epi[\"UMAP 2\"] = np.array(epithelium_colonic_data.obsm[\"X_umap\"])[:, 1]\n",
    "cells_df_col_epi[\"Cell ID\"] = epithelium_colonic_data.obs.index\n",
    "cells_df_col_epi.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "                         \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "cells_df_col_epi = cells_df_col_epi.iloc[:, [0,16,17,18,19,20,21,22,23,24,25,26,27]]\n",
    "\n",
    "# genes dataframe\n",
    "gene_table_col_epi = epithelium_colonic_data.var\n",
    "gene_table_col_epi[\"gene_id\"]=gene_table_col_epi.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ENDOTHELIAL CELLS\n",
    "# cells dataframe \n",
    "cells_df_endo = pd.DataFrame(endothelial_data.obs)\n",
    "\n",
    "# adding the umap data to the cells dataframe\n",
    "#umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "cells_df_endo[\"UMAP 1\"] = np.array(endothelial_data.obsm[\"X_umap\"])[:, 0]\n",
    "cells_df_endo[\"UMAP 2\"] = np.array(endothelial_data.obsm[\"X_umap\"])[:, 1]\n",
    "cells_df_endo[\"Cell ID\"] = endothelial_data.obs.index\n",
    "cells_df_endo.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "                         \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "cells_df_endo = cells_df_endo.iloc[:, [0,16,17,18,19,20,21,22,23,24,25,26,27]]\n",
    "\n",
    "# genes dataframe\n",
    "gene_table_endo = endothelial_data.var\n",
    "gene_table_endo[\"gene_id\"]=gene_table_endo.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "## FIBROBLAST CELLS\n",
    "# cells dataframe \n",
    "cells_df_fib = pd.DataFrame(fib_peri_data.obs)\n",
    "\n",
    "# adding the umap data to the cells dataframe\n",
    "#umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "cells_df_fib[\"UMAP 1\"] = np.array(fib_peri_data.obsm[\"X_umap\"])[:, 0]\n",
    "cells_df_fib[\"UMAP 2\"] = np.array(fib_peri_data.obsm[\"X_umap\"])[:, 1]\n",
    "cells_df_fib[\"Cell ID\"] = fib_peri_data.obs.index\n",
    "cells_df_fib.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "                         \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "cells_df_fib = cells_df_fib.iloc[:, [0,16,17,18,19,20,21,22,23,24,25,26,27]]\n",
    "\n",
    "# genes dataframe\n",
    "gene_table_fib = fib_peri_data.var\n",
    "gene_table_fib[\"gene_id\"]=gene_table_fib.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ILEAL EPITHELIAL CELLS\n",
    "# cells dataframe \n",
    "cells_df_ileal = pd.DataFrame(epi_ileal_data.obs)\n",
    "\n",
    "# adding the umap data to the cells dataframe\n",
    "#umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "cells_df_ileal[\"UMAP 1\"] = np.array(epi_ileal_data.obsm[\"X_umap\"])[:, 0]\n",
    "cells_df_ileal[\"UMAP 2\"] = np.array(epi_ileal_data.obsm[\"X_umap\"])[:, 1]\n",
    "cells_df_ileal[\"Cell ID\"] = epi_ileal_data.obs.index\n",
    "cells_df_ileal.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "                         \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "cells_df_ileal = cells_df_ileal.iloc[:, [0,16,17,18,19,20,21,22,23,24,25,26,27]]\n",
    "\n",
    "# genes dataframe\n",
    "gene_table_ileal = epi_ileal_data.var\n",
    "gene_table_ileal[\"gene_id\"]=gene_table_ileal.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "## INNATE LYMPHOID CELLS\n",
    "# cells dataframe \n",
    "cells_df_ilc = pd.DataFrame(ilc_data.obs)\n",
    "\n",
    "# adding the umap data to the cells dataframe\n",
    "#umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "cells_df_ilc[\"UMAP 1\"] = np.array(ilc_data.obsm[\"X_umap\"])[:, 0]\n",
    "cells_df_ilc[\"UMAP 2\"] = np.array(ilc_data.obsm[\"X_umap\"])[:, 1]\n",
    "cells_df_ilc[\"Cell ID\"] = ilc_data.obs.index\n",
    "cells_df_ilc.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "                         \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "cells_df_ilc = cells_df_ilc.iloc[:, [0,16,17,18,19,20,21,22,23,24,25,26,27]]\n",
    "\n",
    "# genes dataframe\n",
    "gene_table_ilc = ilc_data.var\n",
    "gene_table_ilc[\"gene_id\"]=gene_table_ilc.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "## MYELOID CELLS\n",
    "# cells dataframe \n",
    "cells_df_myel = pd.DataFrame(myeloid_data.obs)\n",
    "\n",
    "# adding the umap data to the cells dataframe\n",
    "#umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "cells_df_myel[\"UMAP 1\"] = np.array(myeloid_data.obsm[\"X_umap\"])[:, 0]\n",
    "cells_df_myel[\"UMAP 2\"] = np.array(myeloid_data.obsm[\"X_umap\"])[:, 1]\n",
    "cells_df_myel[\"Cell ID\"] = myeloid_data.obs.index\n",
    "cells_df_myel.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "                         \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "cells_df_myel = cells_df_myel.iloc[:, [0,16,17,18,19,20,21,22,23,24,25,26,27]]\n",
    "\n",
    "# genes dataframe\n",
    "gene_table_myel = myeloid_data.var\n",
    "gene_table_myel[\"gene_id\"]=gene_table_myel.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "## PLASMA CELLS\n",
    "# cells dataframe \n",
    "cells_df_plasma = pd.DataFrame(plasmacell_data.obs)\n",
    "\n",
    "# adding the umap data to the cells dataframe\n",
    "#umap_np = np.array(bcell_data.obsm[\"X_umap\"])\n",
    "cells_df_plasma[\"UMAP 1\"] = np.array(plasmacell_data.obsm[\"X_umap\"])[:, 0]\n",
    "cells_df_plasma[\"UMAP 2\"] = np.array(plasmacell_data.obsm[\"X_umap\"])[:, 1]\n",
    "cells_df_plasma[\"Cell ID\"] = plasmacell_data.obs.index\n",
    "cells_df_plasma.rename(columns={\"sub_bucket\": \"Cell type\", \"final_analysis\": \"Cell state\", \"MM_scaled\": \"Inflammation score\", \n",
    "                         \"sample_id\": \"Sample ID\"}, inplace=True)\n",
    "cells_df_plasma = cells_df_plasma.iloc[:, [0,16,17,18,19,20,21,22,23,24,25,26,27]]\n",
    "\n",
    "# genes dataframe\n",
    "gene_table_plasma = plasmacell_data.var\n",
    "gene_table_plasma[\"gene_id\"]=gene_table_plasma.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "## B CELLS\n",
    "\n",
    "# Cells charts\n",
    "# creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "stacked_row_chart_bcell = StackedRowChart(\n",
    "    title=\"Abundance plot\",\n",
    "    params=[\"Patient\", \"Cell state\"],\n",
    "    position=[930, 10],\n",
    "    size=[580, 450]\n",
    ")\n",
    "# configuring the row chart\n",
    "stacked_row_chart_bcell.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "stacked_row_chart_bcell.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "\n",
    "#creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "gene_name = \"HSPE1\"\n",
    "\n",
    "dot_plot_bcell = DotPlot(\n",
    "    title= f\"Percent of gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_bcell.index.get_loc(gene_name)}\"],\n",
    "    size=[450, 300],\n",
    "    position=[10, 470]\n",
    ")\n",
    "\n",
    "# configuring the dot plot\n",
    "dot_plot_bcell.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_bcell.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_bcell.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_bcell.set_color_scale(log_scale=False)\n",
    "dot_plot_bcell.set_color_legend(True, [40, 10])\n",
    "dot_plot_bcell.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "box_plot_bcell = BoxPlot(\n",
    "    title=f\"Gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_bcell.index.get_loc(gene_name)}\"],\n",
    "    size=[650, 300],\n",
    "    position=[470, 470]\n",
    ")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot_bcell = ScatterPlot(\n",
    "    title=\"B cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[10, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot_bcell.set_color_by(\"Cell state\")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot2_bcell = ScatterPlot(\n",
    "    title=\"B cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[470, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot2_bcell.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_bcell.index.get_loc(gene_name)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "## CD4T CELLS\n",
    "\n",
    "# Cells charts\n",
    "# creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "stacked_row_chart_cd4t = StackedRowChart(\n",
    "    title=\"Abundance plot\",\n",
    "    params=[\"Patient\", \"Cell state\"],\n",
    "    position=[930, 10],\n",
    "    size=[580, 450]\n",
    ")\n",
    "# configuring the row chart\n",
    "stacked_row_chart_cd4t.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "stacked_row_chart_cd4t.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "#creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "gene_name = \"HSPE1\"\n",
    "\n",
    "dot_plot_cd4t = DotPlot(\n",
    "    title= f\"Percent of gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_cd4t.index.get_loc(gene_name)}\"],\n",
    "    size=[450, 300],\n",
    "    position=[10, 470]\n",
    ")\n",
    "\n",
    "# configuring the dot plot\n",
    "dot_plot_cd4t.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_cd4t.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_cd4t.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_cd4t.set_color_scale(log_scale=False)\n",
    "dot_plot_cd4t.set_color_legend(True, [40, 10])\n",
    "dot_plot_cd4t.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "box_plot_cd4t = BoxPlot(\n",
    "    title=f\"Gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_cd4t.index.get_loc(gene_name)}\"],\n",
    "    size=[650, 300],\n",
    "    position=[470, 470]\n",
    ")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot_cd4t = ScatterPlot(\n",
    "    title=\"CD4+ T cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[10, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot_cd4t.set_color_by(\"Cell state\")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot2_cd4t = ScatterPlot(\n",
    "    title=\"CD4+ T cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[470, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot2_cd4t.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_cd4t.index.get_loc(gene_name)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## cd8T CELLS\n",
    "\n",
    "# Cells charts\n",
    "# creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "stacked_row_chart_cd8t = StackedRowChart(\n",
    "    title=\"Abundance plot\",\n",
    "    params=[\"Patient\", \"Cell state\"],\n",
    "    position=[930, 10],\n",
    "    size=[580, 450]\n",
    ")\n",
    "# configuring the row chart\n",
    "stacked_row_chart_cd8t.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "stacked_row_chart_cd8t.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "\n",
    "#creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "gene_name = \"HSPE1\"\n",
    "\n",
    "dot_plot_cd8t = DotPlot(\n",
    "    title= f\"Percent of gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_cd8t.index.get_loc(gene_name)}\"],\n",
    "    size=[450, 300],\n",
    "    position=[10, 470]\n",
    ")\n",
    "\n",
    "# configuring the dot plot\n",
    "dot_plot_cd8t.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_cd8t.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_cd8t.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_cd8t.set_color_scale(log_scale=False)\n",
    "dot_plot_cd8t.set_color_legend(True, [40, 10])\n",
    "dot_plot_cd8t.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "box_plot_cd8t = BoxPlot(\n",
    "    title=f\"Gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_cd8t.index.get_loc(gene_name)}\"],\n",
    "    size=[650, 300],\n",
    "    position=[470, 470]\n",
    ")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot_cd8t = ScatterPlot(\n",
    "    title=\"CD8+ T innate T NK cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[10, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot_cd8t.set_color_by(\"Cell state\")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot2_cd8t = ScatterPlot(\n",
    "    title=\"CD8+ T innate T NK cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[470, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot2_cd8t.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_cd8t.index.get_loc(gene_name)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "## COLONIC EPITHELIAL CELLS\n",
    "\n",
    "# Cells charts\n",
    "# creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "stacked_row_chart_col_epi = StackedRowChart(\n",
    "    title=\"Abundance plot\",\n",
    "    params=[\"Patient\", \"Cell state\"],\n",
    "    position=[930, 10],\n",
    "    size=[580, 450]\n",
    ")\n",
    "# configuring the row chart\n",
    "stacked_row_chart_col_epi.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "stacked_row_chart_col_epi.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "\n",
    "#creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "gene_name = \"HSPE1\"\n",
    "\n",
    "dot_plot_col_epi = DotPlot(\n",
    "    title= f\"Percent of gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_col_epi.index.get_loc(gene_name)}\"],\n",
    "    size=[450, 300],\n",
    "    position=[10, 470]\n",
    ")\n",
    "\n",
    "# configuring the dot plot\n",
    "dot_plot_col_epi.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_col_epi.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_col_epi.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_col_epi.set_color_scale(log_scale=False)\n",
    "dot_plot_col_epi.set_color_legend(True, [40, 10])\n",
    "dot_plot_col_epi.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "box_plot_col_epi = BoxPlot(\n",
    "    title=f\"Gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_col_epi.index.get_loc(gene_name)}\"],\n",
    "    size=[650, 300],\n",
    "    position=[470, 470]\n",
    ")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot_col_epi = ScatterPlot(\n",
    "    title=\"Colonic epithelial cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[10, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot_col_epi.set_color_by(\"Cell state\")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot2_col_epi = ScatterPlot(\n",
    "    title=\"Colonic epithelial cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[470, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot2_col_epi.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_col_epi.index.get_loc(gene_name)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ENDOTHELIAL CELLS\n",
    "\n",
    "# Cells charts\n",
    "# creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "stacked_row_chart_endo = StackedRowChart(\n",
    "    title=\"Abundance plot\",\n",
    "    params=[\"Patient\", \"Cell state\"],\n",
    "    position=[930, 10],\n",
    "    size=[580, 450]\n",
    ")\n",
    "# configuring the row chart\n",
    "stacked_row_chart_endo.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "stacked_row_chart_endo.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "#creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "gene_name = \"HSPE1\"\n",
    "\n",
    "dot_plot_endo = DotPlot(\n",
    "    title= f\"Percent of gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_endo.index.get_loc(gene_name)}\"],\n",
    "    size=[450, 300],\n",
    "    position=[10, 470]\n",
    ")\n",
    "\n",
    "# configuring the dot plot\n",
    "dot_plot_endo.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_endo.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_endo.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_endo.set_color_scale(log_scale=False)\n",
    "dot_plot_endo.set_color_legend(True, [40, 10])\n",
    "dot_plot_endo.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "box_plot_endo = BoxPlot(\n",
    "    title=f\"Gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_endo.index.get_loc(gene_name)}\"],\n",
    "    size=[650, 300],\n",
    "    position=[470, 470]\n",
    ")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot_endo = ScatterPlot(\n",
    "    title=\"Endothelial cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[10, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot_endo.set_color_by(\"Cell state\")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot2_endo = ScatterPlot(\n",
    "    title=\"Endothelial cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[470, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot2_endo.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_endo.index.get_loc(gene_name)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "## FIBROBLAST PERICYTE CELLS\n",
    "\n",
    "# Cells charts\n",
    "# creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "stacked_row_chart_fib = StackedRowChart(\n",
    "    title=\"Abundance plot\",\n",
    "    params=[\"Patient\", \"Cell state\"],\n",
    "    position=[930, 10],\n",
    "    size=[580, 450]\n",
    ")\n",
    "# configuring the row chart\n",
    "stacked_row_chart_fib.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "stacked_row_chart_fib.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "#creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "gene_name = \"HSPE1\"\n",
    "\n",
    "dot_plot_fib = DotPlot(\n",
    "    title= f\"Percent of gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_fib.index.get_loc(gene_name)}\"],\n",
    "    size=[450, 300],\n",
    "    position=[10, 470]\n",
    ")\n",
    "\n",
    "# configuring the dot plot\n",
    "dot_plot_fib.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_fib.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_fib.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_fib.set_color_scale(log_scale=False)\n",
    "dot_plot_fib.set_color_legend(True, [40, 10])\n",
    "dot_plot_fib.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "box_plot_fib = BoxPlot(\n",
    "    title=f\"Gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_fib.index.get_loc(gene_name)}\"],\n",
    "    size=[650, 300],\n",
    "    position=[470, 470]\n",
    ")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot_fib = ScatterPlot(\n",
    "    title=\"Fibroblast pericyte cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[10, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot_fib.set_color_by(\"Cell state\")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot2_fib = ScatterPlot(\n",
    "    title=\"Fibroblast pericyte cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[470, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot2_fib.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_fib.index.get_loc(gene_name)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ILEAL EPITHELIAL CELLS\n",
    "\n",
    "# Cells charts\n",
    "# creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "stacked_row_chart_ileal = StackedRowChart(\n",
    "    title=\"Abundance plot\",\n",
    "    params=[\"Patient\", \"Cell state\"],\n",
    "    position=[930, 10],\n",
    "    size=[580, 450]\n",
    ")\n",
    "# configuring the row chart\n",
    "stacked_row_chart_ileal.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "stacked_row_chart_ileal.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "#creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "gene_name = \"HSPE1\"\n",
    "\n",
    "dot_plot_ileal = DotPlot(\n",
    "    title= f\"Percent of gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_ileal.index.get_loc(gene_name)}\"],\n",
    "    size=[450, 300],\n",
    "    position=[10, 470]\n",
    ")\n",
    "\n",
    "# configuring the dot plot\n",
    "dot_plot_ileal.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_ileal.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_ileal.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_ileal.set_color_scale(log_scale=False)\n",
    "dot_plot_ileal.set_color_legend(True, [40, 10])\n",
    "dot_plot_ileal.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "box_plot_ileal = BoxPlot(\n",
    "    title=f\"Gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_ileal.index.get_loc(gene_name)}\"],\n",
    "    size=[650, 300],\n",
    "    position=[470, 470]\n",
    ")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot_ileal = ScatterPlot(\n",
    "    title=\"Ileal epithelial cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[10, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot_ileal.set_color_by(\"Cell state\")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot2_ileal = ScatterPlot(\n",
    "    title=\"Ileal epithelial cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[470, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot2_ileal.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_ileal.index.get_loc(gene_name)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "## INNATE LYMPHOID CELLS\n",
    "\n",
    "# Cells charts\n",
    "# creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "stacked_row_chart_ilc = StackedRowChart(\n",
    "    title=\"Abundance plot\",\n",
    "    params=[\"Patient\", \"Cell state\"],\n",
    "    position=[930, 10],\n",
    "    size=[580, 450]\n",
    ")\n",
    "# configuring the row chart\n",
    "stacked_row_chart_ilc.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "stacked_row_chart_ilc.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "#creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "gene_name = \"HSPE1\"\n",
    "\n",
    "dot_plot_ilc = DotPlot(\n",
    "    title= f\"Percent of gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_ilc.index.get_loc(gene_name)}\"],\n",
    "    size=[450, 300],\n",
    "    position=[10, 470]\n",
    ")\n",
    "\n",
    "# configuring the dot plot\n",
    "dot_plot_ilc.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_ilc.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_ilc.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_ilc.set_color_scale(log_scale=False)\n",
    "dot_plot_ilc.set_color_legend(True, [40, 10])\n",
    "dot_plot_ilc.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "box_plot_ilc = BoxPlot(\n",
    "    title=f\"Gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_ilc.index.get_loc(gene_name)}\"],\n",
    "    size=[650, 300],\n",
    "    position=[470, 470]\n",
    ")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot_ilc = ScatterPlot(\n",
    "    title=\"Innate lymphoid cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[10, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot_ilc.set_color_by(\"Cell state\")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot2_ilc = ScatterPlot(\n",
    "    title=\"Innate lymphoid cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[470, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot2_ilc.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_ilc.index.get_loc(gene_name)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "## MYELOID CELLS\n",
    "\n",
    "# Cells charts\n",
    "# creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "stacked_row_chart_myel = StackedRowChart(\n",
    "    title=\"Abundance plot\",\n",
    "    params=[\"Patient\", \"Cell state\"],\n",
    "    position=[930, 10],\n",
    "    size=[580, 450]\n",
    ")\n",
    "# configuring the row chart\n",
    "stacked_row_chart_myel.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "stacked_row_chart_myel.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "#creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "gene_name = \"HSPE1\"\n",
    "\n",
    "dot_plot_myel = DotPlot(\n",
    "    title= f\"Percent of gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_myel.index.get_loc(gene_name)}\"],\n",
    "    size=[450, 300],\n",
    "    position=[10, 470]\n",
    ")\n",
    "\n",
    "# configuring the dot plot\n",
    "dot_plot_myel.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_myel.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_myel.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_myel.set_color_scale(log_scale=False)\n",
    "dot_plot_myel.set_color_legend(True, [40, 10])\n",
    "dot_plot_myel.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "box_plot_myel = BoxPlot(\n",
    "    title=f\"Gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_myel.index.get_loc(gene_name)}\"],\n",
    "    size=[650, 300],\n",
    "    position=[470, 470]\n",
    ")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot_myel = ScatterPlot(\n",
    "    title=\"Myeloid cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[10, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot_myel.set_color_by(\"Cell state\")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot2_myel = ScatterPlot(\n",
    "    title=\"Myeloid cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[470, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot2_myel.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_myel.index.get_loc(gene_name)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "## PLASMA CELLS\n",
    "\n",
    "# Cells charts\n",
    "# creating a row chart based on the Row Chart implementation to show the leiden clustering\n",
    "stacked_row_chart_plasma = StackedRowChart(\n",
    "    title=\"Abundance plot\",\n",
    "    params=[\"Patient\", \"Cell state\"],\n",
    "    position=[930, 10],\n",
    "    size=[580, 450]\n",
    ")\n",
    "# configuring the row chart\n",
    "stacked_row_chart_plasma.set_axis_properties(\"x\", {\"textSize\": 13, \"label\": \"\", \"tickfont\": 10})\n",
    "stacked_row_chart_plasma.set_color_legend(display= True, params=[10,10])\n",
    "\n",
    "#creating a dot plot based on the DotPlot implementation to show the gene expression of selected gene markers\n",
    "gene_name = \"HSPE1\"\n",
    "\n",
    "dot_plot_plasma = DotPlot(\n",
    "    title= f\"Percent of gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_plasma.index.get_loc(gene_name)}\"],\n",
    "    size=[450, 300],\n",
    "    position=[10, 470]\n",
    ")\n",
    "\n",
    "# configuring the dot plot\n",
    "dot_plot_plasma.set_axis_properties(\"x\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_plasma.set_axis_properties(\"y\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_plasma.set_axis_properties(\"ry\", {\"label\": \"\", \"textSize\": 13, \"tickfont\": 10})\n",
    "dot_plot_plasma.set_color_scale(log_scale=False)\n",
    "dot_plot_plasma.set_color_legend(True, [40, 10])\n",
    "dot_plot_plasma.set_fraction_legend(True, [0, 0])\n",
    "\n",
    "\n",
    "# # creating a histogram plot based on the HistogramPlot implementation to show the distribution of the number of genes per counts\n",
    "box_plot_plasma = BoxPlot(\n",
    "    title=f\"Gene expression for {gene_name} per cell state\",\n",
    "    params=[\"Cell state\", f\"Gene expression|{gene_name}(Gene expression)|{gene_table_plasma.index.get_loc(gene_name)}\"],\n",
    "    size=[650, 300],\n",
    "    position=[470, 470]\n",
    ")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot_plasma = ScatterPlot(\n",
    "    title=\"Plasma cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[10, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot_plasma.set_color_by(\"Cell state\")\n",
    "\n",
    "# creating a scatter plot based on the ScatterPlot3D implementation to show the 3 PCA clustering components\n",
    "scatter_plot2_plasma = ScatterPlot(\n",
    "    title=\"Plasma cells\",\n",
    "    params=[\"UMAP 1\", \"UMAP 2\"],\n",
    "    size=[450, 450],\n",
    "    position=[470, 10],\n",
    "    default_color=\"#377eb8\",\n",
    "    brush=\"default\",\n",
    "    on_filter=\"hide\",\n",
    "    radius=5,\n",
    "    opacity=0.8,\n",
    ")\n",
    "\n",
    "#Â configuring the scatter plot\n",
    "scatter_plot2_plasma.set_color_by(f\"Gene expression|{gene_name}(Gene expression)|{gene_table_plasma.index.get_loc(gene_name)}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set up and serve the MDV project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# setting up and serving the MDV project\n",
    "base = os.path.expanduser('~/mdv')\n",
    "project_path = os.path.join(base, 'taurus') # defining the location where the project metadata will be stored\n",
    "p = MDVProject(os.path.expanduser(project_path), delete_existing=True)\n",
    "\n",
    "# adding the two data sources to the project\n",
    "# B cells\n",
    "p.add_datasource(\"B cells\", cells_df_bcell)\n",
    "p.add_datasource(\"B genes\", gene_table_bcell)\n",
    "# CD4T cells\n",
    "p.add_datasource(\"CD4T cells\", cells_df_cd4t)\n",
    "p.add_datasource(\"CD4T genes\", gene_table_cd4t)\n",
    "# CD8T cells\n",
    "p.add_datasource(\"CD8T cells\", cells_df_cd8t)\n",
    "p.add_datasource(\"CD8T genes\", gene_table_cd8t)\n",
    "# Colonic epithelial cells\n",
    "#p.add_datasource(\"Colonic epithelial cells\", cells_df_col_epi)\n",
    "#p.add_datasource(\"Colonic epithelial genes\", gene_table_col_epi)\n",
    "# Endothelial cells\n",
    "p.add_datasource(\"Endothelial cells\", cells_df_endo)\n",
    "p.add_datasource(\"Endothelial genes\", gene_table_endo)\n",
    "# Endothelial cells\n",
    "p.add_datasource(\"Fibroblast pericyte cells\", cells_df_fib)\n",
    "p.add_datasource(\"Fibroblast pericyte genes\", gene_table_fib)\n",
    "# Ileal epithelial cells\n",
    "p.add_datasource(\"Ileal epithelial cells\", cells_df_ileal)\n",
    "p.add_datasource(\"Ileal epithelial genes\", gene_table_ileal)\n",
    "# Innate lymphoid cells\n",
    "p.add_datasource(\"Innate lymphoid cells\", cells_df_ilc)\n",
    "p.add_datasource(\"Innate lymphoid genes\", gene_table_ilc)\n",
    "# Myeloid cells\n",
    "p.add_datasource(\"Myeloid cells\", cells_df_myel)\n",
    "p.add_datasource(\"Myeloid genes\", gene_table_myel)\n",
    "# Plasma cells\n",
    "p.add_datasource(\"Plasma cells\", cells_df_plasma)\n",
    "p.add_datasource(\"Plasma genes\", gene_table_plasma)\n",
    "\n",
    "# creating the link between the two datasets so that selecting a subset of genes to add the expression in cells is enabled\n",
    "p.add_rows_as_columns_link(\"B cells\",\"B genes\",\"gene_id\",\"Gene expression\")\n",
    "p.add_rows_as_columns_subgroup(\"B cells\",\"B genes\",\"Gene expression\",bcell_data.layers[\"counts\"].toarray()) #add the gene expression \n",
    "\n",
    "p.add_rows_as_columns_link(\"CD4T cells\",\"CD4T genes\",\"gene_id\",\"Gene expression\")\n",
    "p.add_rows_as_columns_subgroup(\"CD4T cells\",\"CD4T genes\",\"Gene expression\",cd4tcell_data.layers[\"counts\"].toarray()) #add the gene expression \n",
    "\n",
    "p.add_rows_as_columns_link(\"CD8T cells\",\"CD8T genes\",\"gene_id\",\"Gene expression\")\n",
    "p.add_rows_as_columns_subgroup(\"CD8T cells\",\"CD8T genes\",\"Gene expression\",cd8tcell_data.layers[\"counts\"].toarray()) #add the gene expression\n",
    "\n",
    "# p.add_rows_as_columns_link(\"Colonic epithelial cells\",\"Colonic epithelial genes\",\"gene_id\",\"Gene expression\")\n",
    "# p.add_rows_as_columns_subgroup(\"Colonic epithelial cells\",\"Colonic epithelial genes\",\"Gene expression\",epithelium_colonic_data.layers[\"counts\"].toarray()) #add the gene expression\n",
    "\n",
    "p.add_rows_as_columns_link(\"Endothelial cells\",\"Endothelial genes\",\"gene_id\",\"Gene expression\")\n",
    "p.add_rows_as_columns_subgroup(\"Endothelial cells\",\"Endothelial genes\",\"Gene expression\",endothelial_data.layers[\"counts\"].toarray()) #add the gene expression\n",
    "\n",
    "p.add_rows_as_columns_link(\"Fibroblast pericyte cells\",\"Fibroblast pericyte genes\",\"gene_id\",\"Gene expression\")\n",
    "p.add_rows_as_columns_subgroup(\"Fibroblast pericyte cells\",\"Fibroblast pericyte genes\",\"Gene expression\",fib_peri_data.layers[\"counts\"].toarray()) #add the gene expression\n",
    "\n",
    "p.add_rows_as_columns_link(\"Ileal epithelial cells\",\"Ileal epithelial genes\",\"gene_id\",\"Gene expression\")\n",
    "p.add_rows_as_columns_subgroup(\"Ileal epithelial cells\",\"Ileal epithelial genes\",\"Gene expression\",epi_ileal_data.layers[\"counts\"].toarray()) #add the gene expression\n",
    "\n",
    "p.add_rows_as_columns_link(\"Innate lymphoid cells\",\"Innate lymphoid genes\",\"gene_id\",\"Gene expression\")\n",
    "p.add_rows_as_columns_subgroup(\"Innate lymphoid cells\",\"Innate lymphoid genes\",\"Gene expression\",ilc_data.layers[\"counts\"].toarray()) #add the gene expression\n",
    "\n",
    "p.add_rows_as_columns_link(\"Myeloid cells\",\"Myeloid genes\",\"gene_id\",\"Gene expression\")\n",
    "p.add_rows_as_columns_subgroup(\"Myeloid cells\",\"Myeloid genes\",\"Gene expression\",myeloid_data.layers[\"counts\"].toarray()) #add the gene expression\n",
    "\n",
    "p.add_rows_as_columns_link(\"Plasma cells\",\"Plasma genes\",\"gene_id\",\"Gene expression\")\n",
    "p.add_rows_as_columns_subgroup(\"Plasma cells\",\"Plasma genes\",\"Gene expression\",plasmacell_data.layers[\"counts\"].toarray()) #add the gene expression\n",
    "\n",
    "# converting the chart implementation outputs to JSON and setting up the project view\n",
    "list_charts_cells_bcell = []\n",
    "list_charts_cells_cd4t = []\n",
    "list_charts_cells_cd8t = []\n",
    "list_charts_cells_col_epi = []\n",
    "list_charts_cells_endo = []\n",
    "list_charts_cells_fib = []\n",
    "list_charts_cells_ileal = []\n",
    "list_charts_cells_ilc = []\n",
    "list_charts_cells_myel = []\n",
    "list_charts_cells_plasma = []\n",
    "\n",
    "# cells panel\n",
    "list_charts_cells_bcell.extend([stacked_row_chart_bcell.plot_data, dot_plot_bcell.plot_data, box_plot_bcell.plot_data, \n",
    "                                scatter_plot_bcell.plot_data, scatter_plot2_bcell.plot_data])\n",
    "list_charts_cells_cd4t.extend([stacked_row_chart_cd4t.plot_data, dot_plot_cd4t.plot_data, box_plot_cd4t.plot_data, \n",
    "                                scatter_plot_cd4t.plot_data, scatter_plot2_cd4t.plot_data])\n",
    "list_charts_cells_cd8t.extend([stacked_row_chart_cd8t.plot_data, dot_plot_cd8t.plot_data, box_plot_cd8t.plot_data, \n",
    "                                scatter_plot_cd8t.plot_data, scatter_plot2_cd8t.plot_data])\n",
    "list_charts_cells_col_epi.extend([stacked_row_chart_col_epi.plot_data, dot_plot_col_epi.plot_data, box_plot_col_epi.plot_data, \n",
    "                                scatter_plot_col_epi.plot_data, scatter_plot2_col_epi.plot_data])\n",
    "list_charts_cells_endo.extend([stacked_row_chart_endo.plot_data, dot_plot_endo.plot_data, box_plot_endo.plot_data, \n",
    "                                scatter_plot_endo.plot_data, scatter_plot2_endo.plot_data])\n",
    "list_charts_cells_fib.extend([stacked_row_chart_fib.plot_data, dot_plot_fib.plot_data, box_plot_fib.plot_data, \n",
    "                                scatter_plot_fib.plot_data, scatter_plot2_fib.plot_data])\n",
    "list_charts_cells_ileal.extend([stacked_row_chart_ileal.plot_data, dot_plot_ileal.plot_data, box_plot_ileal.plot_data, \n",
    "                                scatter_plot_ileal.plot_data, scatter_plot2_ileal.plot_data])\n",
    "list_charts_cells_ilc.extend([stacked_row_chart_ilc.plot_data, dot_plot_ilc.plot_data, box_plot_ilc.plot_data, \n",
    "                                scatter_plot_ilc.plot_data, scatter_plot2_ilc.plot_data])\n",
    "list_charts_cells_myel.extend([stacked_row_chart_myel.plot_data, dot_plot_myel.plot_data, box_plot_myel.plot_data, \n",
    "                                scatter_plot_myel.plot_data, scatter_plot2_myel.plot_data])\n",
    "list_charts_cells_plasma.extend([stacked_row_chart_plasma.plot_data, dot_plot_plasma.plot_data, box_plot_plasma.plot_data, \n",
    "                                scatter_plot_plasma.plot_data, scatter_plot2_plasma.plot_data])\n",
    "\n",
    "# setting the config combining the two panels\n",
    "view_config_bcell = {'initialCharts': {\"B cells\": list_charts_cells_bcell}}\n",
    "view_config_cd4t = {'initialCharts': {\"CD4T cells\": list_charts_cells_cd4t}}\n",
    "view_config_cd8t = {'initialCharts': {\"CD8T cells\": list_charts_cells_cd8t}}\n",
    "view_config_col_epi = {'initialCharts': {\"Colonic epithelial cells\": list_charts_cells_col_epi}}\n",
    "view_config_endo = {'initialCharts': {\"Endothelial cells\": list_charts_cells_endo}}\n",
    "view_config_fib = {'initialCharts': {\"Fibroblast pericyte cells\": list_charts_cells_fib}}\n",
    "view_config_ileal = {'initialCharts': {\"Ileal epithelial cells\": list_charts_cells_ileal}}\n",
    "view_config_ilc = {'initialCharts': {\"Innate lymphoid cells\": list_charts_cells_ilc}}\n",
    "view_config_myel = {'initialCharts': {\"Myeloid cells\": list_charts_cells_myel}}\n",
    "view_config_plasma = {'initialCharts': {\"Plasma cells\": list_charts_cells_plasma}}\n",
    "\n",
    "# adding the view to the project configuration\n",
    "p.set_view(\"B cells\", view_config_bcell)\n",
    "p.set_view(\"CD4T cells\", view_config_cd4t)\n",
    "p.set_view(\"CD8T cells\", view_config_cd8t)\n",
    "#p.set_view(\"Colonic epithelial cells\", view_config_col_epi)\n",
    "p.set_view(\"Endothelial cells\", view_config_endo)\n",
    "p.set_view(\"Fibroblast pericyte cells\", view_config_fib)\n",
    "p.set_view(\"Ileal epithelial cells\", view_config_ileal)\n",
    "p.set_view(\"Innate lymphoid cells\", view_config_ilc)\n",
    "p.set_view(\"Myeloid cells\", view_config_myel)\n",
    "p.set_view(\"Plasma cells\", view_config_plasma)\n",
    "\n",
    "p.set_editable(True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# setting keys and variables\n",
      "# Crawl the local repository to get a list of relevant file paths\n",
      "Block 'b1: Local repo crawling' took 0.0473 seconds\n",
      "# Initialize a text splitter for chunking the code strings\n",
      "# Split the code documents into chunks using the text splitter\n",
      "Block 'b2: Text splitter initialising' took 0.0009 seconds\n",
      "# Initialize an instance of the OpenAIEmbeddings class\n",
      "Block 'b3: Embeddings creating' took 0.0298 seconds\n",
      "client=<openai.resources.embeddings.Embeddings object at 0x309e809b0> async_client=<openai.resources.embeddings.AsyncEmbeddings object at 0x30ae822d0> model='text-embedding-3-large' dimensions=None deployment='text-embedding-ada-002' openai_api_version=None openai_api_base=None openai_api_type=None openai_proxy=None embedding_ctx_length=8191 openai_api_key=SecretStr('**********') openai_organization=None allowed_special=None disallowed_special=None chunk_size=1000 max_retries=2 request_timeout=None headers=None tiktoken_enabled=True tiktoken_model_name=None show_progress_bar=False model_kwargs={} skip_empty=False default_headers=None default_query=None retry_min_seconds=4 retry_max_seconds=20 http_client=None http_async_client=None check_embedding_ctx_length=True\n",
      "# Create an index from the embedded code chunks\n",
      "# Use FAISS (Facebook AI Similarity Search) to create a searchable index\n",
      "Block 'b4: FAISS database creating' took 1.7054 seconds\n",
      "# Initialize the retriever from the FAISS index\n",
      "Block 'b5: Retriever creating' took 0.0000 seconds\n",
      "created Flask <Flask 'mdvtools.server'>\n",
      "socketio initialized\n",
      "websocket/chat enabled for \n",
      " * Serving Flask app 'mdvtools.server'\n",
      " * Debug mode: on\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.\n",
      " * Running on all addresses (0.0.0.0)\n",
      " * Running on http://127.0.0.1:5050\n",
      " * Running on http://129.67.46.69:5050\n",
      "Press CTRL+C to quit\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET / HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recieved request to project_index\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/js/mdv.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/mdv.css HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/bot-message-square-4f2KPFIG.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/csvWorker-e8cMUEqV.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /datasources.json HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /state.json HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /views.json HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"POST /get_view HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/img/fa-solid-900.woff2 HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:56:13] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing state response. websocket: True\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [13/Nov/2024 15:57:44] \"POST /get_view HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:57:44] \"POST /get_data HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:57:44] \"POST /get_data HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:57:44] \"POST /get_data HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:57:44] \"GET /static/assets/catWorker-D5lhlpxE.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:57:44] \"GET /static/assets/catColWorker-BRK20MZB.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:57:44] \"GET /static/assets/binWorker-DIC2QqPk.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:57:44] \"GET /static/assets/binWorker-DIC2QqPk.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:57:44] \"GET /static/assets/boxPlotWorker-BvOiGWqU.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:57:44] \"GET /static/assets/binWorker-DIC2QqPk.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:57:47] \"GET /static/img/roboto-latin-400-normal.woff2 HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:57:54] \"GET /datasources.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:57:55] \"GET /views.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:57:55] \"GET /state.json HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:57:55] \"GET /chat_log.json HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing state response. websocket: True\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [13/Nov/2024 15:58:15] \"POST /save_state HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:17] \"POST /get_view HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:17] \"POST /get_data HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:17] \"POST /get_data HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:17] \"POST /get_data HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:17] \"GET /static/assets/catWorker-D5lhlpxE.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:17] \"GET /static/assets/binWorker-DIC2QqPk.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:18] \"GET /static/assets/binWorker-DIC2QqPk.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:18] \"GET /static/assets/boxPlotWorker-BvOiGWqU.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:18] \"GET /static/assets/binWorker-DIC2QqPk.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:18] \"GET /static/assets/catColWorker-BRK20MZB.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:20] \"GET /?view=B+cells HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:20] \"GET /static/assets/mdv.css HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:20] \"GET /static/js/mdv.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:20] \"GET /static/assets/bot-message-square-4f2KPFIG.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/csvWorker-e8cMUEqV.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /datasources.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /state.json HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recieved request to project_index\n",
      "processing state response. websocket: True\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /views.json HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"POST /get_view HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/img/fa-solid-900.woff2 HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"POST /get_data HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"POST /get_data HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"POST /get_data HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/catWorker-D5lhlpxE.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/catColWorker-BRK20MZB.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/binWorker-DIC2QqPk.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/filteredIndexWorker-CEl1713S.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/binWorker-DIC2QqPk.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/boxPlotWorker-BvOiGWqU.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:21] \"GET /static/assets/binWorker-DIC2QqPk.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:23] \"GET /datasources.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:23] \"GET /views.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:23] \"GET /state.json HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:23] \"GET /chat_log.json HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [13/Nov/2024 15:58:23] \"GET /static/img/roboto-latin-400-normal.woff2 HTTP/1.1\" 304 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing state response. websocket: True\n"
     ]
    }
   ],
   "source": [
    "# serving the project\n",
    "p.serve()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venvjupyter",
   "language": "python",
   "name": "venvjupyter"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
